# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [6.0.0](https://github.com/c3g/GenPipes/compare/5.1.0...6.0.0)

### Merged

- update longread docs and remove bitbucket refs [`#20`](https://github.com/c3g/GenPipes/pull/20)
- Dev [`#19`](https://github.com/c3g/GenPipes/pull/19)
- methylseq : pass dedup bam to gembs_call [`#18`](https://github.com/c3g/GenPipes/pull/18)
- Djerba template updates [`#17`](https://github.com/c3g/GenPipes/pull/17)
- Rnaseq single end [`#16`](https://github.com/c3g/GenPipes/pull/16)
- make ini name consistent with pipeline name [`#15`](https://github.com/c3g/GenPipes/pull/15)
- Longread dnaseq [`#10`](https://github.com/c3g/GenPipes/pull/10)
- Gz chip [`#8`](https://github.com/c3g/GenPipes/pull/8)
- Longread dnaseq [`#6`](https://github.com/c3g/GenPipes/pull/6)
- Version bump to 5.1.0 [`#561`](https://github.com/c3g/GenPipes/pull/561)

### Commits

- First changes to rename utils script into tools and include them within genpipes like genpipes tools ... [`a22b137`](https://github.com/c3g/GenPipes/commit/a22b1377e56a9e20f73f68a35c6fbe000be6ad66)
- Linting [`2d19f40`](https://github.com/c3g/GenPipes/commit/2d19f409c7dabd39ab267d59fb1dbbadcb1cc292)
- Adding design/readset validator + updating to 2025 [`ee3b957`](https://github.com/c3g/GenPipes/commit/ee3b9578f0be22331b6490c052efe7d746b74fe1)
- adding longread pipeline with revio and nanopore protocols [`46fb859`](https://github.com/c3g/GenPipes/commit/46fb8598e078b19e5cbf0c46e1a0818d9d0a3c8f)
- Consolidating log_report to rely on new prologue and epilogue only [`b6f7b7f`](https://github.com/c3g/GenPipes/commit/b6f7b7ff3292a7fc806ae368dd0be4bda897a455)
- Removing json option and creation [`5dcac71`](https://github.com/c3g/GenPipes/commit/5dcac71945f5bd02f2892d007e7d5305eb691f13)
- remove nanopore pipeline, now in longread_dnaseq [`ef38866`](https://github.com/c3g/GenPipes/commit/ef3886621eccdfa49432b28bdb61bc34be4bef28)
- Removing old 2 log_reports into 1 single [`00fb114`](https://github.com/c3g/GenPipes/commit/00fb1140d420f5381a184fd8880be8b95701e538)
- Consolidating autocomplate and shtab + README + adding autocomplete to dist [`5d349ee`](https://github.com/c3g/GenPipes/commit/5d349ee14af1334e05caa90314372c0279c63aa5)
- updating input to input_file in bash.ln calls [`96e920d`](https://github.com/c3g/GenPipes/commit/96e920df71bfd25869eaa00f5f9096b91b0b14f5)

## [5.1.0](https://github.com/c3g/GenPipes/compare/5.0.3...5.1.0) - 2025-02-03

### Merged

- Release 5.1.0 [`#560`](https://github.com/c3g/GenPipes/pull/560)
- addressing Melissas report for batch mode [`#559`](https://github.com/c3g/GenPipes/pull/559)
- Dnaseq multiple normals [`#558`](https://github.com/c3g/GenPipes/pull/558)
- Version bump to 5.0.3 [`#549`](https://github.com/c3g/GenPipes/pull/549)

### Commits

- fix indentation in cnvkit_batch step loop [`9d9b396`](https://github.com/c3g/GenPipes/commit/9d9b3961f6f8db5fa5d42b3f7c335922276443b8)
- dnaseq : remove different alignment directory structure when multiple normals [`1e33940`](https://github.com/c3g/GenPipes/commit/1e33940e07d81df35cf6eaeac5b2287e5825e2de)
- updating README for release [`32fb1fa`](https://github.com/c3g/GenPipes/commit/32fb1fa0f92df6f5a93311c1d6a2aebabf89cb34)
- Add error if tumor pair names are not unique [`580484c`](https://github.com/c3g/GenPipes/commit/580484cc721a77eca52f477e0130d7162ac1b776)
- Version bump to 5.1.0 [`9dae02c`](https://github.com/c3g/GenPipes/commit/9dae02c3db41267b2317952246b8afcc6d596d80)
- Version bump to 5.0.4-beta [`ddb6125`](https://github.com/c3g/GenPipes/commit/ddb6125f25040ae9180143a249dd86329048a5f0)
- Improve error message if tumor pair names are not unique [`357dc9f`](https://github.com/c3g/GenPipes/commit/357dc9f3e86a89381b2e052791771ded06844cb4)
- Merged tumor_dnaseq_merge into dev [`772f3a9`](https://github.com/c3g/GenPipes/commit/772f3a9a9325a6234c36840c0678a8cba83b8134)

## [5.0.3](https://github.com/c3g/GenPipes/compare/5.0.2...5.0.3) - 2024-12-09

### Merged

- Release 5.0.3 [`#548`](https://github.com/c3g/GenPipes/pull/548)
- bfx : remove stray space in conpair wrapper [`#547`](https://github.com/c3g/GenPipes/pull/547)
- Version bump to 5.0.2 [`#545`](https://github.com/c3g/GenPipes/pull/545)

### Commits

- update README [`ee62cca`](https://github.com/c3g/GenPipes/commit/ee62cca8fbe0e2e949142832c3ff1468d09c8f8a)
- update README [`5c3a8e6`](https://github.com/c3g/GenPipes/commit/5c3a8e642557ee21bf67566a762b0570de33bb62)
- fix invalid escape sequence [`7df1d4c`](https://github.com/c3g/GenPipes/commit/7df1d4c0e30c0cf5b18995475fb9cc303d30bbaa)
- Version bump to 5.0.3 [`277fd12`](https://github.com/c3g/GenPipes/commit/277fd123862e385a93d7a6c8dc31b389ccab97db)
- Version bump to 5.0.3-beta [`1bc1c6c`](https://github.com/c3g/GenPipes/commit/1bc1c6cbb11af20e9b9b8dad4cd1597bccb6d0be)

## [5.0.2](https://github.com/c3g/GenPipes/compare/5.0.1...5.0.2) - 2024-12-05

### Merged

- Release 5.0.2 [`#544`](https://github.com/c3g/GenPipes/pull/544)
- GenPipes chipseq : remove homer report job [`#542`](https://github.com/c3g/GenPipes/pull/542)
- Chipseq fix [`#541`](https://github.com/c3g/GenPipes/pull/541)
- Pcgr grep fix [`#535`](https://github.com/c3g/GenPipes/pull/535)
- Version bump to 5.0.1 [`#533`](https://github.com/c3g/GenPipes/pull/533)

### Commits

- GenPipes chipseq : rm ihec metrics from MultiQC [`fd0b283`](https://github.com/c3g/GenPipes/commit/fd0b28386dea6f275c5a3b2e2588b26a85725ee8)
- GenPipes chipseq : add sample metrics tsv back [`d43004c`](https://github.com/c3g/GenPipes/commit/d43004c31bf38869c8fc1e1be6361666a7598587)
- GenPipes : updates for pcgr 2.1.0 cpsr inputs [`5eb110b`](https://github.com/c3g/GenPipes/commit/5eb110b653e9a8fc317c8d67b69101e83eab9694)
- GenPipes chipseq : fix loop for merging ihec metrics to include all inputs [`61cbc8f`](https://github.com/c3g/GenPipes/commit/61cbc8fdd0c19df4631b291d2b2b257c1a5acce2)
- update README for release [`8c5f8c9`](https://github.com/c3g/GenPipes/commit/8c5f8c9af19bc9e74083219b62b568dacfdf6f36)
- GenPipes chipseq : add nmb peaks and frip to SampleMetrics.tsv [`9a5879a`](https://github.com/c3g/GenPipes/commit/9a5879ad9f2697953f5b925604153a1075eceba7)
- GenPipes chipseq : metrics report debug [`862a56e`](https://github.com/c3g/GenPipes/commit/862a56e9ecf09814c721a9558a41b13655927798)
- GenPipes chipseq : fix loop for merging ihec metrics to include all inputs [`ac92d21`](https://github.com/c3g/GenPipes/commit/ac92d217dd03bb96790841134abfae681ed9f606)
- Updates chunking script to fix missing indents [`d6e05a4`](https://github.com/c3g/GenPipes/commit/d6e05a41e89f17bfadbacc81d0eee00622409a87)
- GenPipes rnaseq : increase cit resources for new pcgr version [`c2d8f50`](https://github.com/c3g/GenPipes/commit/c2d8f50a7b2933df425e1a922ec13fe5a6097635)

## [5.0.1](https://github.com/c3g/GenPipes/compare/5.0.0...5.0.1) - 2024-09-24

### Merged

- Release 5.0.1 [`#532`](https://github.com/c3g/GenPipes/pull/532)
- Autocomplete updates [`#531`](https://github.com/c3g/GenPipes/pull/531)
- Version bump to 5.0.0 [`#530`](https://github.com/c3g/GenPipes/pull/530)

### Commits

- update autocompletes [`06275f7`](https://github.com/c3g/GenPipes/commit/06275f77be19418475058512a7557c0aafb5d9b5)
- update READMEs [`be6efc5`](https://github.com/c3g/GenPipes/commit/be6efc5f580d32ff44f27d4e1213678f9850ff96)
- update paths in README for new directory structure [`b40d70a`](https://github.com/c3g/GenPipes/commit/b40d70a9b3f2774fff6f0173d52a8d1c6ee98ed6)
- update MUGQIC_PIPELINES_HOME variable to GENPIPES_INIS in the context of accessing config file [`c05ff72`](https://github.com/c3g/GenPipes/commit/c05ff7201669b369ce85520b8b7f465dc5a6cc75)
- preparations for release [`42c5bd7`](https://github.com/c3g/GenPipes/commit/42c5bd73b6f421d9566a8ce507fa662c71b3660f)
- Version bump to 5.0.1-beta [`6d25df8`](https://github.com/c3g/GenPipes/commit/6d25df89ef5b589f0998ae13fc6c09966660eab8)
- add autocomplete files to pip package [`8128541`](https://github.com/c3g/GenPipes/commit/812854127b289cf9419e51ba69d3392a0a281bd1)

## [5.0.0](https://github.com/c3g/GenPipes/compare/4.6.1...5.0.0) - 2024-09-23

### Merged

- Release 5.0.0 [`#529`](https://github.com/c3g/GenPipes/pull/529)
- Ref: #268 update schema for pipelines v5.0 [`#528`](https://github.com/c3g/GenPipes/pull/528)
- Version fixes [`#527`](https://github.com/c3g/GenPipes/pull/527)
- add logging used for debugging [`#526`](https://github.com/c3g/GenPipes/pull/526)
- Submit debug [`#525`](https://github.com/c3g/GenPipes/pull/525)
- Submit debug [`#524`](https://github.com/c3g/GenPipes/pull/524)
- MethylseqGemBS [`#523`](https://github.com/c3g/GenPipes/pull/523)
- Package [`#522`](https://github.com/c3g/GenPipes/pull/522)
- Package [`#521`](https://github.com/c3g/GenPipes/pull/521)
- GenPipes dnaseq : genome fasta for sv protocol cit [`#520`](https://github.com/c3g/GenPipes/pull/520)
- Package [`#519`](https://github.com/c3g/GenPipes/pull/519)
- Package [`#518`](https://github.com/c3g/GenPipes/pull/518)
- Package [`#517`](https://github.com/c3g/GenPipes/pull/517)
- fix samtools_sort other options, remove setting output format explicitely [`#516`](https://github.com/c3g/GenPipes/pull/516)
- Package [`#515`](https://github.com/c3g/GenPipes/pull/515)
- Package [`#514`](https://github.com/c3g/GenPipes/pull/514)
- Tumor dnaseq merge [`#513`](https://github.com/c3g/GenPipes/pull/513)
- Tumor dnaseq merge [`#512`](https://github.com/c3g/GenPipes/pull/512)
- Tumor dnaseq merge [`#510`](https://github.com/c3g/GenPipes/pull/510)
- Tumor dnaseq merge [`#487`](https://github.com/c3g/GenPipes/pull/487)
- Version bump to 4.6.1 [`#508`](https://github.com/c3g/GenPipes/pull/508)

### Commits

- resources : update workflow schemas and rm deprecated ones [`3f98ab0`](https://github.com/c3g/GenPipes/commit/3f98ab0f7c7700b0818dc4763a44286c77778d5e)
- rebase package branch with dev, resolve conflicts [`f54bdd1`](https://github.com/c3g/GenPipes/commit/f54bdd113e55d7ff4f55266c5dfc39bf9b603d21)
- rebasing with dev prior to merge [`898571f`](https://github.com/c3g/GenPipes/commit/898571f8673477fcd3d56229c3c65a9438d1d7e8)
- merge optimizations for cnvkit, cpsr, pcgr, improved comments, changes to get property for protocols [`bcfe3d0`](https://github.com/c3g/GenPipes/commit/bcfe3d030c9bd9065a563821eea9a60053a83a7a)
- update READMEs [`32b9d70`](https://github.com/c3g/GenPipes/commit/32b9d7020efcb942020be178dd160a9631b42841)
- GenPipes ampliconseq : deprecate qiime protocol [`495b092`](https://github.com/c3g/GenPipes/commit/495b0924dd13c997404e916202d9073f3c49c6f1)
- Cit fixes, remove deprecated protocol (mpileup), steps and deprecated steps/modules and update module versions [`adb57f3`](https://github.com/c3g/GenPipes/commit/adb57f39e388f4c3266504a7d3bbecf2153802a4)
- Trying a new docstring + reformatting dnaseq pipeline [`a95c1f4`](https://github.com/c3g/GenPipes/commit/a95c1f4aa4a3dd1f3ebcc20fb0c5ac37eb75c86b)
- GenPipes hicseq : deprecate HiCseq pipeline [`3c9d687`](https://github.com/c3g/GenPipes/commit/3c9d6874da06b2a8839fe07c08654fc2cb70b60a)
- convert to fstrings here applicable, combine varscan2 steps for fastpass and ensemble protocols [`40387be`](https://github.com/c3g/GenPipes/commit/40387be3bedaa38bac55be0322b009fe713e950a)

## [4.6.1](https://github.com/c3g/GenPipes/compare/4.6.0...4.6.1) - 2024-05-21

### Merged

- Release 4.6.1 [`#507`](https://github.com/c3g/GenPipes/pull/507)
- Version bump to 4.6.0 [`#496`](https://github.com/c3g/GenPipes/pull/496)

### Commits

- Version bump to 4.6.1 [`89cac35`](https://github.com/c3g/GenPipes/commit/89cac35f18f57ef005d33bc7336012f377caa761)
- GenPipes dnaseq : added missing definition of strelka2_bed_file [`19ae086`](https://github.com/c3g/GenPipes/commit/19ae08638afc718f275146855f45e4d6d90bbade)

## [4.6.0](https://github.com/c3g/GenPipes/compare/4.5.0...4.6.0) - 2024-05-01

### Merged

- Release 4.6.0 [`#495`](https://github.com/c3g/GenPipes/pull/495)
- GenPipes covseq : resources for cit [`#494`](https://github.com/c3g/GenPipes/pull/494)
- Pcgr name [`#493`](https://github.com/c3g/GenPipes/pull/493)
- Pcgr name [`#492`](https://github.com/c3g/GenPipes/pull/492)
- Rocky fix [`#490`](https://github.com/c3g/GenPipes/pull/490)
- Rocky fix [`#489`](https://github.com/c3g/GenPipes/pull/489)
- GenPipes tumor_pair : keep paired_ensemble, cobalt, and amber directories when generating clean commands [`#488`](https://github.com/c3g/GenPipes/pull/488)
- GenPipes : add zipping of bisSNP output vcfs [`#485`](https://github.com/c3g/GenPipes/pull/485)
- HotFix dev [`#484`](https://github.com/c3g/GenPipes/pull/484)
- Version bump to 4.5.0 [`#483`](https://github.com/c3g/GenPipes/pull/483)

### Commits

- GenPipes tumor_pair : pcgr output file symlinks for long sample names [`cc5cb7b`](https://github.com/c3g/GenPipes/commit/cc5cb7b54215e33d200483ada55014bbda04a61b)
- GenPipes rnaseq : pcgr output file links for long sample names [`702579a`](https://github.com/c3g/GenPipes/commit/702579abe641001a42e481dff9051240f81f66e7)
- Add ini for cardinal [`56518aa`](https://github.com/c3g/GenPipes/commit/56518aa4347942d83f8e6aab9c6f74c309c4bd0c)
- Preparations for release [`1427f2f`](https://github.com/c3g/GenPipes/commit/1427f2f14b032e10b91a60748feaa4e770bdcfe5)
- GenPipes : replace zless with zcat [`4f6781f`](https://github.com/c3g/GenPipes/commit/4f6781f78571d339aa2961355290404d2e8b1ee8)
- GenPipes ampliconseq : prevent clashing python modules [`256b1bf`](https://github.com/c3g/GenPipes/commit/256b1bfa32d525fe169074bef9720dfcaa4a8cee)
- Escaping tab symbol to be resolved by bash and not python [`ccf9876`](https://github.com/c3g/GenPipes/commit/ccf987672fb334c9536a24891db8f2dae80ff9fc)
- GenPipes ampliconseq : debug rarefaction [`cb1784d`](https://github.com/c3g/GenPipes/commit/cb1784d2e35f4e65f46d587afaff69572a6db573)
- GenPipes methylseq : revert bisSNP version to 0.82.2 due to corrupted vcfs from v.1.0.0 [`08bc518`](https://github.com/c3g/GenPipes/commit/08bc5186aebd649660f73a7da2803bc9d381bd38)
- GenPipes tumor_pair : pcgr report for samples with long names [`1ac5cc3`](https://github.com/c3g/GenPipes/commit/1ac5cc3d751716fa068c9a180a4a349025a635a0)

## [4.5.0](https://github.com/c3g/GenPipes/compare/4.4.5...4.5.0) - 2024-02-09

### Merged

- Release 4.5.0 [`#482`](https://github.com/c3g/GenPipes/pull/482)
- GenPipes dnaseq : use snpeff genome named in ini, instead of hardcoding hg19 [`#481`](https://github.com/c3g/GenPipes/pull/481)
- Debug typo [`#480`](https://github.com/c3g/GenPipes/pull/480)
- README.md edited online with Bitbucket: umask update [`#478`](https://github.com/c3g/GenPipes/pull/478)
- Debugging typo in error message [`#479`](https://github.com/c3g/GenPipes/pull/479)
- HotFix dev [`#475`](https://github.com/c3g/GenPipes/pull/475)
- core scheduler : add attribute check when creating job_list name [`#474`](https://github.com/c3g/GenPipes/pull/474)
- Methylseq fix [`#473`](https://github.com/c3g/GenPipes/pull/473)
- CollectWGSmetrics [`#472`](https://github.com/c3g/GenPipes/pull/472)
- Pt deliverables metrics [`#471`](https://github.com/c3g/GenPipes/pull/471)
- Version bump to 4.4.5 [`#469`](https://github.com/c3g/GenPipes/pull/469)

### Commits

- Cleaning + init test [`436b341`](https://github.com/c3g/GenPipes/commit/436b34108fecb9e7846454de8830678d14249dea)
- GenPipes tumor_pair : add CollectWgsMetrics [`52e489d`](https://github.com/c3g/GenPipes/commit/52e489d1e761ae5f6d5bf3faf59ce7a947f8005b)
- GenPipes tumor_pair : add CollectWgsMetrics [`cb70b70`](https://github.com/c3g/GenPipes/commit/cb70b700f28bce330f740407817af5ea03833c55)
- Updating gitignore + trailing white spaces + adding back modules explicitly loaded [`dfa4b94`](https://github.com/c3g/GenPipes/commit/dfa4b9446f756d95f51f14886df5b0c7bccd7b67)
- GenPipes tumor_pair : fixing qualimap multiqc report [`2ed967b`](https://github.com/c3g/GenPipes/commit/2ed967b5d557e2ddb597ec296a1ab856aa3e8a38)
- GenPipes tumor_pair : fixing qualimap multiqc report [`ee31318`](https://github.com/c3g/GenPipes/commit/ee31318c14938a9064204a575b741bc26ee184e6)
- Testing importing all [`9de18d5`](https://github.com/c3g/GenPipes/commit/9de18d5f576ad8d411d598530b6ec925009965e2)
- GenPipe tumor_pair : add CollectWgsMetrics options, cit resources [`5687354`](https://github.com/c3g/GenPipes/commit/5687354f8d2051157711f9e58da12ae01a83d4e2)
- GenPipe tumor_pair : add CollectWgsMetrics options, cit resources [`4fed799`](https://github.com/c3g/GenPipes/commit/4fed7991e1fe5c4671ebd6ba46fc6475a242d527)
- Version bump to 4.5.0 [`56df351`](https://github.com/c3g/GenPipes/commit/56df351fd9a1db0a24b74a1d66421d39e0edafb6)

## [4.4.5](https://github.com/c3g/GenPipes/compare/4.4.4...4.4.5) - 2023-11-14

### Merged

- Release 4.4.5 [`#468`](https://github.com/c3g/GenPipes/pull/468)
- GenPipes common : add if/else in step 5 merge step [`#467`](https://github.com/c3g/GenPipes/pull/467)
- Tp step6 fix [`#466`](https://github.com/c3g/GenPipes/pull/466)
- GenPipes cit : gemini and strelka fixes [`#465`](https://github.com/c3g/GenPipes/pull/465)
- Merge link fix [`#463`](https://github.com/c3g/GenPipes/pull/463)
- GenPipes cit : resources for gemini [`#464`](https://github.com/c3g/GenPipes/pull/464)
- GenPipes : cit resources [`#462`](https://github.com/c3g/GenPipes/pull/462)
- Htseq fix [`#460`](https://github.com/c3g/GenPipes/pull/460)
- Loading and unloading python when calling job2json_project_tracking withing GenPipes [`#459`](https://github.com/c3g/GenPipes/pull/459)
- DiffBind th [`#457`](https://github.com/c3g/GenPipes/pull/457)
- Cit sortmerna fix [`#458`](https://github.com/c3g/GenPipes/pull/458)
- Try debug for re-ordering modules (among others) [`#456`](https://github.com/c3g/GenPipes/pull/456)
- Cit chr19 [`#455`](https://github.com/c3g/GenPipes/pull/455)
- GZ rnaseq sortmeRNA [`#454`](https://github.com/c3g/GenPipes/pull/454)
- Project tracking json [`#453`](https://github.com/c3g/GenPipes/pull/453)
- Adding rnaseqc2 to stringtie section of rnaseq.py [`#434`](https://github.com/c3g/GenPipes/pull/434)
- Version bump to 4.4.4 [`#452`](https://github.com/c3g/GenPipes/pull/452)

### Commits

- Adding Rnaseq light and sv metrics for MoH [`60d1250`](https://github.com/c3g/GenPipes/commit/60d12503e4d8ec0fe4dc9ad5c2e7c85bbe404b54)
- chipseq - merge_and_call_individual_gvcf - Consolidating gatk4 bfx with ini_section to make sure resources are well linked [`ede194a`](https://github.com/c3g/GenPipes/commit/ede194a4225a704268a020577711cba0e8896f28)
- Adding project tracking json in use in tumor pair [`c71223a`](https://github.com/c3g/GenPipes/commit/c71223ac19e04ea968d68462ed21ec00d0fbdb08)
- job2json_project_tracking [`761145d`](https://github.com/c3g/GenPipes/commit/761145d7cb9da7d533d13978a3a681233792f975)
- Version bump to 4.4.5 [`a0957ff`](https://github.com/c3g/GenPipes/commit/a0957ffafd3d458c179a246f94ac6cd6b1b4c88d)
- Adding json project tracking dedicated args + parsing conpair metrics + job2json_pt bfx for updating metrics [`50d5fbd`](https://github.com/c3g/GenPipes/commit/50d5fbdf44aa5bf3c7109d82a8e1fea5a395baab)
- commit after checking out dev & squashing [`405f28c`](https://github.com/c3g/GenPipes/commit/405f28c998d207549ce5caf09ff3d87c79e0cf43)
- Debug [`cd1de74`](https://github.com/c3g/GenPipes/commit/cd1de741369ab84c2df5448d8a5b2f419a044eb6)
- Debug [`3830b62`](https://github.com/c3g/GenPipes/commit/3830b62ef6615a23feecae0596f1bc575fd7541c)
- Addition of sortmerna 1 [`9fb360b`](https://github.com/c3g/GenPipes/commit/9fb360bf03ec998996671ed577b7b771d2b3e948)

## [4.4.4](https://github.com/c3g/GenPipes/compare/4.4.3...4.4.4) - 2023-09-08

### Merged

- Release 4.4.4 [`#451`](https://github.com/c3g/GenPipes/pull/451)
- GenPipes tumor_pair : interval_list job name and dependency issues [`#448`](https://github.com/c3g/GenPipes/pull/448)
- Kallisto index renaming [`#449`](https://github.com/c3g/GenPipes/pull/449)
- GenPipes : change job name for interval list creation in recalibration step [`#447`](https://github.com/c3g/GenPipes/pull/447)
- Version bump to 4.4.3 [`#445`](https://github.com/c3g/GenPipes/pull/445)

### Commits

- Merged dev into mutect_fix [`a854557`](https://github.com/c3g/GenPipes/commit/a8545574fc47d486a7966a4a16e6a32b5b536a80)
- rnaseq_light - kallisto - index renaming to match kallisto version [`49f00cc`](https://github.com/c3g/GenPipes/commit/49f00cc734ad5b83a4ffc66e566d6ed88cfc3509)
- Version bump to 4.4.4 [`fa9ea1e`](https://github.com/c3g/GenPipes/commit/fa9ea1e66da509b381fe89843036daebbcdef0d6)
- rnaseq_light - kallisto - Upgrading kallisto to latest + increasing resources for cit [`47db2c7`](https://github.com/c3g/GenPipes/commit/47db2c7cbb1909c62b07909fff2f7af3e609671b)
- VERSION bump to 4.4.4.-beta [`6e3217b`](https://github.com/c3g/GenPipes/commit/6e3217b8956cc6b3b4666f5dd9fff7aa0e8e2f96)
- Merged master into dev [`7d0f747`](https://github.com/c3g/GenPipes/commit/7d0f747f18b984341e6bf095b8d34ddc2f5eb760)

## [4.4.3](https://github.com/c3g/GenPipes/compare/4.4.2...4.4.3) - 2023-08-23

### Merged

- Release 4.4.3 [`#444`](https://github.com/c3g/GenPipes/pull/444)
- Cit fixes [`#443`](https://github.com/c3g/GenPipes/pull/443)
- MOH permissions [`#436`](https://github.com/c3g/GenPipes/pull/436)
- Readme improvements [`#442`](https://github.com/c3g/GenPipes/pull/442)
- Prepare table fix [`#441`](https://github.com/c3g/GenPipes/pull/441)
- Star fusion restart [`#439`](https://github.com/c3g/GenPipes/pull/439)
- GenPipes tumor_pair : add md5 for recalibration output bam [`#440`](https://github.com/c3g/GenPipes/pull/440)
- Wiggle fix [`#438`](https://github.com/c3g/GenPipes/pull/438)
- Minimap2 align check [`#437`](https://github.com/c3g/GenPipes/pull/437)
- Improve cit inis [`#435`](https://github.com/c3g/GenPipes/pull/435)
- Improve cit inis [`#433`](https://github.com/c3g/GenPipes/pull/433)
- Rnaseq moh multiqc [`#432`](https://github.com/c3g/GenPipes/pull/432)
- Version bump to 4.4.2 [`#431`](https://github.com/c3g/GenPipes/pull/431)

### Commits

- update pipeline READMEs prior to release [`14242b6`](https://github.com/c3g/GenPipes/commit/14242b6bdeb7d995985dd926073d14775a8754e4)
- GenPipes : updating documentation with Shaloos suggestions [`16a7cc8`](https://github.com/c3g/GenPipes/commit/16a7cc8a667ca9a959b71d513840db9f80bbd320)
- GenPipes : cit ini updates for beluga [`fcbbd6e`](https://github.com/c3g/GenPipes/commit/fcbbd6ec6264a85de2e72d5250b7f796bf6384e6)
- bfx : add samtools quickcheck [`7d9e01a`](https://github.com/c3g/GenPipes/commit/7d9e01a0cc3c18b54dc74cc6e796303feb91d045)
- update README [`08a6d65`](https://github.com/c3g/GenPipes/commit/08a6d65e4756dffc6d5bc143eb8735a6864869f5)
- GenPipes cit : adjust mem and walltime [`3389502`](https://github.com/c3g/GenPipes/commit/3389502c0592f22ea023e809cd66a5e8bb31f2f7)
- GenPipes cit : adjustments to fix oom and timeouts [`ca2dd54`](https://github.com/c3g/GenPipes/commit/ca2dd54de4d5bdfe3c3063367add177fe05a638d)
- GenPipes rnaseq : cit walltime fixes [`de5b7b8`](https://github.com/c3g/GenPipes/commit/de5b7b84f2802679d9c7685b1258c0ddfbb5c0ef)
- version bump to 4.4.3 [`de07ec6`](https://github.com/c3g/GenPipes/commit/de07ec668c81846c0ba57f231392ad8725fa8d69)
- Merged dev into cit_fixes [`eac2d6a`](https://github.com/c3g/GenPipes/commit/eac2d6a048026bf46105ce6b613ba8dc9e6294ec)

## [4.4.2](https://github.com/c3g/GenPipes/compare/4.4.1...4.4.2) - 2023-06-21

### Merged

- Release 4.4.2 [`#430`](https://github.com/c3g/GenPipes/pull/430)
- Methylseq multiqc [`#429`](https://github.com/c3g/GenPipes/pull/429)
- GenPipes : add symlink for trimmomatic log to use with multiqc [`#428`](https://github.com/c3g/GenPipes/pull/428)
- Multiqc cit fix [`#427`](https://github.com/c3g/GenPipes/pull/427)
- Rm hard clip [`#426`](https://github.com/c3g/GenPipes/pull/426)
- Cit multiqc fixes [`#424`](https://github.com/c3g/GenPipes/pull/424)
- Rnaseq multiqc [`#423`](https://github.com/c3g/GenPipes/pull/423)
- GenPipes - updated dates in licence [`#422`](https://github.com/c3g/GenPipes/pull/422)
- GenPipes tumor_pair : add remove command before bcbio_ensembl call to avoid silent error when output exists [`#421`](https://github.com/c3g/GenPipes/pull/421)
- changed the cpulimit [`#420`](https://github.com/c3g/GenPipes/pull/420)
- Chipseq blacklist filter [`#418`](https://github.com/c3g/GenPipes/pull/418)
- GenPipes : improved README for dnaseq pipeline [`#417`](https://github.com/c3g/GenPipes/pull/417)
- common.py edited online with Bitbucket - fix for sambamba merge dependency [`#416`](https://github.com/c3g/GenPipes/pull/416)
- Sambamba merge fix [`#415`](https://github.com/c3g/GenPipes/pull/415)
- Tumor Pair - Fixed gripss calls [`#413`](https://github.com/c3g/GenPipes/pull/413)
- Version bump to 4.4.1 [`#412`](https://github.com/c3g/GenPipes/pull/412)

### Commits

- Tumor Pair - Updated multiqc prep steps and job [`352863a`](https://github.com/c3g/GenPipes/commit/352863a1cb9828c4dda8cad24a0a24336580fb83)
- DNASeq - Fixed multiqc step dependencies [`8172a2a`](https://github.com/c3g/GenPipes/commit/8172a2ad57f764e7a8c7336d6f20106b5263ab2f)
- GenPipes methylseq : add multiqc to pipeline [`f2b9ccb`](https://github.com/c3g/GenPipes/commit/f2b9ccb09dc62c09f11d504e406bfd9626075a2c)
- GenPipes : methylseq - ihec table to multiqc format [`8961e31`](https://github.com/c3g/GenPipes/commit/8961e31d986f7f339578051f2e49b9dc5d4732c0)
- update pipeline READMEs prior to release [`8888ad0`](https://github.com/c3g/GenPipes/commit/8888ad03ef433274ffd4b4dc8e755ed060e60304)
- GenPipes tumor_pair : loop over report_files output for multiqc symlinks [`a0b2122`](https://github.com/c3g/GenPipes/commit/a0b21222a0f455e5d9b42163782629c83b6c38e4)
- GenPipes rnaseq : rm bam_hard_clip step [`7eb52d4`](https://github.com/c3g/GenPipes/commit/7eb52d4521f3a74fd86f56b0aea29a5e463710c3)
- Utils - added dump_ChangeLog.sh in utils [`cf60298`](https://github.com/c3g/GenPipes/commit/cf60298ec840c107ff1c3d0a7aa636965a985655)
- Genpipes tumor_pair : create multiqc report dir as part of jobs [`0eae1f9`](https://github.com/c3g/GenPipes/commit/0eae1f99c153d1d49353dcdc06dce4ad7b189309)
- GenPipes : rm sequencing_center refs to wrong centre, add underscores to McGill_Genome_Centre [`271b96e`](https://github.com/c3g/GenPipes/commit/271b96ecc86569d312b6a1653ec6685c1432dc8a)

## [4.4.1](https://github.com/c3g/GenPipes/compare/4.4.0...4.4.1) - 2023-03-14

### Merged

- Release 4.4.1 [`#411`](https://github.com/c3g/GenPipes/pull/411)
- Version bump to 4.4.0 [`#410`](https://github.com/c3g/GenPipes/pull/410)

### Commits

- Resources - adding 'gget' install script and updating others [`07f45a2`](https://github.com/c3g/GenPipes/commit/07f45a23fed899fc576ae1835cdba8bbceb8c3c3)
- Resources - adding / updating install scripts [`087fd9b`](https://github.com/c3g/GenPipes/commit/087fd9b3fc663e73d1bfc51ea93fc6a4b8398881)
- GenPipes - in prep for bug-fix release 4.4.1 [`412c2e3`](https://github.com/c3g/GenPipes/commit/412c2e3d69f6e42eea4cec346a4a0da4622add2d)
- EpiQC - correcting chromimpute_preprocess command : do not use "rm" and force symlink creation with the "-f" flag [`6a5222f`](https://github.com/c3g/GenPipes/commit/6a5222f2deb424aba3e1c0e0931c0559e5962273)
- Resources - update mosdepth version to 0.3.3 in install script [`c27edf1`](https://github.com/c3g/GenPipes/commit/c27edf154938edcbcf91cfa1ec5a49ece846afc2)
- Version bump to 4.4.1-beta [`ed8e359`](https://github.com/c3g/GenPipes/commit/ed8e359e6232221b18aa5f53997924a022eb3ec5)
- RNASeq Light - Added mkdir to the kalliso command [`252b6ee`](https://github.com/c3g/GenPipes/commit/252b6ee50e6548be61c6e6e70c316c6702940230)

## [4.4.0](https://github.com/c3g/GenPipes/compare/4.3.2...4.4.0) - 2023-03-09

### Merged

- Release 4.4.0 [`#409`](https://github.com/c3g/GenPipes/pull/409)
- Rnaseq cancer : update genome version used for cpsr and pcgr [`#408`](https://github.com/c3g/GenPipes/pull/408)
- Cit fixes mcj [`#407`](https://github.com/c3g/GenPipes/pull/407)
- methylseq : do not require methylation_protocol and mapping_implementation parameters to be set for hybrid protocol [`#406`](https://github.com/c3g/GenPipes/pull/406)
- Cit fixes mcj [`#405`](https://github.com/c3g/GenPipes/pull/405)
- EpiBrain methylseq [`#404`](https://github.com/c3g/GenPipes/pull/404)
- Cit fixes mcj [`#403`](https://github.com/c3g/GenPipes/pull/403)
- Rebase rnaseq variant - variant and cancer protocols [`#370`](https://github.com/c3g/GenPipes/pull/370)
- Tpsv [`#401`](https://github.com/c3g/GenPipes/pull/401)
- Tp fix multiqc [`#399`](https://github.com/c3g/GenPipes/pull/399)
- issue caused by qiime_catenate loading two python modules [`#398`](https://github.com/c3g/GenPipes/pull/398)
- ability to skip 2-pass star alignment and only do 1 pass [`#394`](https://github.com/c3g/GenPipes/pull/394)
- gatk_indel_realigner : added two cd commands [`#393`](https://github.com/c3g/GenPipes/pull/393)
- Gerardo rnaseq bigwig [`#390`](https://github.com/c3g/GenPipes/pull/390)
- DNASEQ - ensemble_metasv dependency fix [`#391`](https://github.com/c3g/GenPipes/pull/391)
- Version bump to 4.3.2 [`#389`](https://github.com/c3g/GenPipes/pull/389)

### Commits

- GenPipes - updating READMEs before release [`cf28289`](https://github.com/c3g/GenPipes/commit/cf28289151d7d457ef03ce2857cf2cc0bffdd140)
- Resources - Adding some new install scripts [`70d27f7`](https://github.com/c3g/GenPipes/commit/70d27f7c171083aafbc7cd572b45c8efe7a9bdb5)
- RNASeq Light - revamping the call to Kallisto to have per gene counts instead of per readset [`91bca69`](https://github.com/c3g/GenPipes/commit/91bca69e855d6277c39f6b4f85da1094aea1c364)
- Resources - Adding AGeNT install script and updating script for VarDict [`79b74d0`](https://github.com/c3g/GenPipes/commit/79b74d038d2f0dba2d2e8f74e55b0ba5d8866e6a)
- Resources - updated version in some install scripts [`699683c`](https://github.com/c3g/GenPipes/commit/699683caee757943126d99c9a094c35de4e8978d)
- Resources - Adding install script for hmmcopy-utils and ichorCNA [`d83071d`](https://github.com/c3g/GenPipes/commit/d83071de7d055f158f7a4b86cd68e34f5c1c5b33)
- EpiQC - Fixing `chromimpute_convert` dependency with `chromimpute_preprocess` [`658004c`](https://github.com/c3g/GenPipes/commit/658004c6ef02602844ee4e21ae8ec6380101535f)
- gatk_indel_realigner : set -fixMisencodedQuals flag only when QualityOffsets are 64 (solexa/illumina1.5+) [`4cab100`](https://github.com/c3g/GenPipes/commit/4cab1006f7d0462142a517a88a5b8e6361c0af60)
- EpiQC - fixed file path : distinction between files for job dependency and files for direct access [`6f4b9a9`](https://github.com/c3g/GenPipes/commit/6f4b9a9bc98bfea7ab1a1da9262b376e3e732d02)
- RNASeq Light - Adding kallisto wrapper in bfx [`ffd223d`](https://github.com/c3g/GenPipes/commit/ffd223d030e59bde6c5dc77795e46d18de640784)

## [4.3.2](https://github.com/c3g/GenPipes/compare/4.3.1...4.3.2) - 2022-12-08

### Merged

- Release 4.3.2 [`#388`](https://github.com/c3g/GenPipes/pull/388)
- GenPipes fixes and format after cit - all pipelines [`#385`](https://github.com/c3g/GenPipes/pull/385)
- HotFix dev [`#387`](https://github.com/c3g/GenPipes/pull/387)
- Removing index from sorting as the bai is generated by sorting [`#386`](https://github.com/c3g/GenPipes/pull/386)
- Bwa sambamba splitting [`#384`](https://github.com/c3g/GenPipes/pull/384)
- Core - Scheduler : corrected `memory` call in `cpu` from dev [`#383`](https://github.com/c3g/GenPipes/pull/383)
- Version bump to 4.3.1 [`#382`](https://github.com/c3g/GenPipes/pull/382)

### Commits

- GenPipes - DEV : updating CHANGELOG and setting beta version [`df5aa27`](https://github.com/c3g/GenPipes/commit/df5aa276cff533b48be174c1c63121b7ff17020a)
- Tumor Pair - Improving the restart mecanisms [`3341cbc`](https://github.com/c3g/GenPipes/commit/3341cbc889461ebe716050d4aea19a9f2ed2a18c)
- Corrections after cit + formating [`40cf403`](https://github.com/c3g/GenPipes/commit/40cf4033b5b6187ce80e1914f656afe5762645b9)
- Some more fixes induced by cit [`53e3e05`](https://github.com/c3g/GenPipes/commit/53e3e05fc026c1af67aad33a75e840129e7e0517)
- GenPipes - code cleaning and formating [`90b9535`](https://github.com/c3g/GenPipes/commit/90b953538d423951837001b3288e786650cc4b9c)
- Fixes after cit + code standardization [`2bebb47`](https://github.com/c3g/GenPipes/commit/2bebb47ef91ec7e11abc8c3f07bd058a2b8e5c6a)
- HicSeq - working on restart/resume mechanisms + doing some reformating [`10727fe`](https://github.com/c3g/GenPipes/commit/10727fed4ac2a019f97a084b41407aae4196a325)
- GenPipes - improved averall restart/resume mechanisims with relative path [`ef89a8f`](https://github.com/c3g/GenPipes/commit/ef89a8f43855fe83e3e65a1c8b232ff7bcb62568)
- GenPipes - Fixes after cit on Béluga [`185ed83`](https://github.com/c3g/GenPipes/commit/185ed83100113d109aff873f27ea35582ac4fb19)
- Install script - updates in module install scripts [`20c0951`](https://github.com/c3g/GenPipes/commit/20c0951e3e82fdeee3c0a2c283a7d2fb7c51bdad)

## [4.3.1](https://github.com/c3g/GenPipes/compare/4.3.0...4.3.1) - 2022-10-04

### Merged

- Release 4.3.1 [`#381`](https://github.com/c3g/GenPipes/pull/381)
- Remove explicit request for memory in pbs call when --force_mem_per_cpu option is used [`#380`](https://github.com/c3g/GenPipes/pull/380)
- HotFix dev [`#378`](https://github.com/c3g/GenPipes/pull/378)
- Tumor pair hotfix [`#375`](https://github.com/c3g/GenPipes/pull/375)
- Add an option to fail jobs if a regex pattern is found in the logs [`#376`](https://github.com/c3g/GenPipes/pull/376)
- Add timestamp to trace config filename [`#374`](https://github.com/c3g/GenPipes/pull/374)
- nanopore_covseq.base.ini edited to re-add the GPU_QUEUE value. Without it the pipeline crashes on Abacus. [`#373`](https://github.com/c3g/GenPipes/pull/373)
- MetaSV Delly Bug [`#372`](https://github.com/c3g/GenPipes/pull/372)
- added differential methylation to the methylseq pipeline [`#371`](https://github.com/c3g/GenPipes/pull/371)
- Move sample readset and design to core from bfx [`#338`](https://github.com/c3g/GenPipes/pull/338)
- tumor_pair.py edited online with Bitbucket [`#369`](https://github.com/c3g/GenPipes/pull/369)
- Epiqc code improvement [`#368`](https://github.com/c3g/GenPipes/pull/368)

### Commits

- modified the methylseq/cit.ini [`3908f5e`](https://github.com/c3g/GenPipes/commit/3908f5e0f8cbdbe3248c4a19100bbd17f526b1bf)
- hicseq - General - Improving hicup job and removing path hardcoded in ini [`42245c7`](https://github.com/c3g/GenPipes/commit/42245c79ea4e026ddd060e5bfd28ab91ccf83ea4)
- chnaged resources in cit.ini [`3166cdf`](https://github.com/c3g/GenPipes/commit/3166cdf1ba166bc04de8c13376f7bb03ac386d26)
- GenPipes : updating READMEs for release [`759bbfb`](https://github.com/c3g/GenPipes/commit/759bbfb9fb71cb83caf70ca1436291877951c1eb)
- hicseq - General - Improving the way genome assembly name is handled [`d837f8d`](https://github.com/c3g/GenPipes/commit/d837f8dae18ca374159dd9cec78fc0eb7ed8db33)
- hicseq - hicrep - Code cleaning [`0c1b223`](https://github.com/c3g/GenPipes/commit/0c1b223c251b391b2f72cfca0a3d62728aa0ce37)
- hicseq - hicup - No longer creating folder and removing things in the step [`3018b34`](https://github.com/c3g/GenPipes/commit/3018b34ffdf221fca1300da52f6d90372c191ef9)
- hicseq - quasar_qc - Allowing more than 1 "_" in chr name [`1619b24`](https://github.com/c3g/GenPipes/commit/1619b24b1ccfbcd518bd696ce9af1bce62001c29)
- move sample readset and design to core from bfx [`92049a7`](https://github.com/c3g/GenPipes/commit/92049a7776a712820ffee17530f033b8bd7d7ed5)
- add a fail_on_pattern ini option [`a74d954`](https://github.com/c3g/GenPipes/commit/a74d9547ca3f583e8cf7bbaa09c506df17275108)

## [4.3.0](https://github.com/c3g/GenPipes/compare/4.2.1...4.3.0) - 2022-06-15

### Merged

- Release 4.3.0.1 [`#367`](https://github.com/c3g/GenPipes/pull/367)
- Release 4.3.0.1 [`#366`](https://github.com/c3g/GenPipes/pull/366)
- Updating READMEs [`#365`](https://github.com/c3g/GenPipes/pull/365)
- Version bump to 4.3.0 [`#364`](https://github.com/c3g/GenPipes/pull/364)
- Release 4.3.0 [`#363`](https://github.com/c3g/GenPipes/pull/363)
- modified readme file with dragen update [`#362`](https://github.com/c3g/GenPipes/pull/362)
- Version bump to 4.2.1 [`#359`](https://github.com/c3g/GenPipes/pull/359)

### Commits

- Updating CHANGELOG from master after release [`8636775`](https://github.com/c3g/GenPipes/commit/8636775733f2f8a45ec75299befb4655d45766fc)
- General - General - Formalizing README doc. [`3079952`](https://github.com/c3g/GenPipes/commit/307995288346c5ead9b9d9bc5febd4be38dc9e4d)
- GenPipes - prep for new release [`ec90b7a`](https://github.com/c3g/GenPipes/commit/ec90b7abee39053636964d94531812cc35eedf84)
- Added multiline comments to methylseq pipeline [`b772f84`](https://github.com/c3g/GenPipes/commit/b772f84a6abcc58b4b9c3d9320d1144fea695a05)
- Version bump to 4.3.0 - for real [`5766788`](https://github.com/c3g/GenPipes/commit/57667889e8b64c3b5b54373b1b8ec05314c94f9e)
- fixed some issues in epiQC md file [`9684c5e`](https://github.com/c3g/GenPipes/commit/9684c5e2f448194d36946f2cc9f350b8d39c78ef)
- rnaseq - General - Updating doc. [`0c39884`](https://github.com/c3g/GenPipes/commit/0c39884cba2084c55872e3ecd68f78b34a5c7fde)
- modified readme file [`486907a`](https://github.com/c3g/GenPipes/commit/486907ab6ab1748c65c0016cac07cc791e00f1ba)

## [4.2.1](https://github.com/c3g/GenPipes/compare/4.2.0...4.2.1) - 2022-06-09

### Merged

- Release 4.2.1 [`#358`](https://github.com/c3g/GenPipes/pull/358)
- General - General - Fixing UnboundLocalError: local variable 'job_name_prefix' referenced before assignment [`#357`](https://github.com/c3g/GenPipes/pull/357)
- HotFix dev [`#356`](https://github.com/c3g/GenPipes/pull/356)
- Soft jsondb gsoc2020 [`#181`](https://github.com/c3g/GenPipes/pull/181)
- HotFix dev [`#350`](https://github.com/c3g/GenPipes/pull/350)
- Unmapped reads [`#355`](https://github.com/c3g/GenPipes/pull/355)
- Methylation Dragen Pipeline [`#346`](https://github.com/c3g/GenPipes/pull/346)
- Version bump to 4.2.0 [`#354`](https://github.com/c3g/GenPipes/pull/354)

### Commits

- 1. Pushing pre-final code [`810321f`](https://github.com/c3g/GenPipes/commit/810321f0785a3d9f3c896a3e7b989b9717870d86)
- added comments and fix hybrid mode single pass directional-complement [`d82bab0`](https://github.com/c3g/GenPipes/commit/d82bab0123e1aff4026d5d3aa0204dba8d6100ed)
- added split_dragen_methylation_report [`454abe6`](https://github.com/c3g/GenPipes/commit/454abe68008a613a7409ee13185c943a32bef1a6)
- Update readme and version for release [`b092002`](https://github.com/c3g/GenPipes/commit/b09200275e1aea413a1d7873ffb03bf57d34fb56)
- Migration to argparse [`8d223ca`](https://github.com/c3g/GenPipes/commit/8d223caba113a182982bf0c4996c3e077b2a37d0)
- fixing issues with the done file tracking system [`dfa349d`](https://github.com/c3g/GenPipes/commit/dfa349d6336bbc4c2d75d1c4bd0420b809d1a79c)
- fixed issues in output file tracking [`5949d7b`](https://github.com/c3g/GenPipes/commit/5949d7bc5d0bb7851f27c6ac686b1d3270000e55)
- GenPipes - C3G Software Stack : Added sending of the combined JSON to url + reformating the module helpers [`3303a59`](https://github.com/c3g/GenPipes/commit/3303a59fc2a2ef1059f74450665b1893e22aa7ed)
- Some documentation edits, README updates, fixed bad variable naming [`a51bba0`](https://github.com/c3g/GenPipes/commit/a51bba0f14f4e06436b396301770b3fb4d98b552)
- Implement Modified dependency of dragen command to allo mv commands [`8ed0ab8`](https://github.com/c3g/GenPipes/commit/8ed0ab89b1f2d73aec9cf82303a65eb033ecfd73)

## [4.2.0](https://github.com/c3g/GenPipes/compare/4.1.3...4.2.0) - 2022-06-01

### Merged

- Generating READMEs for 4.2.0 [`#353`](https://github.com/c3g/GenPipes/pull/353)
- Version bump to 4.1.3 [`#352`](https://github.com/c3g/GenPipes/pull/352)

## [4.1.3](https://github.com/c3g/GenPipes/compare/4.1.2...4.1.3) - 2022-05-31

### Merged

- Release 4.1.3 [`#351`](https://github.com/c3g/GenPipes/pull/351)
- Tp reports [`#348`](https://github.com/c3g/GenPipes/pull/348)
- Tp reports [`#347`](https://github.com/c3g/GenPipes/pull/347)
- covseq - prepare_report - Adding pandoc from cvmfs [`#341`](https://github.com/c3g/GenPipes/pull/341)
- HotFix dev [`#342`](https://github.com/c3g/GenPipes/pull/342)
- GenPipes - Tumor Pair : set the output_dirs global variable to better handle outputs, thus dependencies [`#339`](https://github.com/c3g/GenPipes/pull/339)
- General - ini - Standardizing ini base section [`#340`](https://github.com/c3g/GenPipes/pull/340)
- covseq - ini - Fixing bam2fq number of threads hard coded [`#336`](https://github.com/c3g/GenPipes/pull/336)
- changed job name for merge_gatk [`#335`](https://github.com/c3g/GenPipes/pull/335)
- chipseq - macs2_callpeak + homer_annotate - making genomre_size not required in ini [`#334`](https://github.com/c3g/GenPipes/pull/334)
- Chip snp pubudu [`#324`](https://github.com/c3g/GenPipes/pull/324)
- chipseq - base.ini - Fixing cluster_cpu old way [`#330`](https://github.com/c3g/GenPipes/pull/330)
- Tumorpair hotfix strelka [`#331`](https://github.com/c3g/GenPipes/pull/331)
- Nanopore_covseq + General [`#325`](https://github.com/c3g/GenPipes/pull/325)
- Dnaseqsv fix eh [`#329`](https://github.com/c3g/GenPipes/pull/329)
- dnaseq - General - HotFix [`#328`](https://github.com/c3g/GenPipes/pull/328)
- changed mugqic_tools version to version 2.10 [`#327`](https://github.com/c3g/GenPipes/pull/327)
- Chipseq urgent fix [`#326`](https://github.com/c3g/GenPipes/pull/326)
- Fix dnaseqsv eh [`#320`](https://github.com/c3g/GenPipes/pull/320)
- chipseq - macs2_callpeak - Fixing typo error [`#322`](https://github.com/c3g/GenPipes/pull/322)
- Adding cpulimit for ncovtools (on Abacus only) [`#321`](https://github.com/c3g/GenPipes/pull/321)
- new ini file setup with a single common ini for different clusters [`#319`](https://github.com/c3g/GenPipes/pull/319)
- GenPipes - DNASeq SV : fixed delly call [`#317`](https://github.com/c3g/GenPipes/pull/317)
- Final adjustment to the nanopore_covseq base ini to optimize job submission on Abacus. [`#316`](https://github.com/c3g/GenPipes/pull/316)
- chipseq - macs2_callpeak - Fixing other_options to be used via ini [`#315`](https://github.com/c3g/GenPipes/pull/315)
- Version bump to 4.1.2 [`#314`](https://github.com/c3g/GenPipes/pull/314)

### Commits

- GenPipes - Tumor Pair : fixed dependency issues by using relative path everywhere in the pipeline, no more absolute path. Also setting 'samples' for all the jobs. [`70c9b89`](https://github.com/c3g/GenPipes/commit/70c9b89af357c5e3a07324174897061c31fcdeda)
- adding cpsr/pcgr reporting with addition of manta, cnvkit to ensemble protocol.  Step clean up as well [`35e5c27`](https://github.com/c3g/GenPipes/commit/35e5c27fc90638511c0dc92d40068b9b441988ca)
- Added variant calling using GATK4 to chipseq pipeline [`d5189ed`](https://github.com/c3g/GenPipes/commit/d5189ed7ed3d67fcbc34e32a525ede60302b12bf)
- Added variant calling using GATK4 to chipseq pipeline [`896157b`](https://github.com/c3g/GenPipes/commit/896157b1f2eb5644e17a5ce0d225f3616f8e86ea)
- Added variant calling using GATK4 to chipseq pipeline [`a40ed3a`](https://github.com/c3g/GenPipes/commit/a40ed3afa3c038f180b67ef1db396fdbadc33bc6)
- Added variant calling using GATK4 to chipseq pipeline [`f595d71`](https://github.com/c3g/GenPipes/commit/f595d7196177728eff91ff72880e5e75145ce48a)
- fixed a bug in input file list of haplotypecaller [`c5d8d3e`](https://github.com/c3g/GenPipes/commit/c5d8d3e9247a4be3387a909d65362dc28e5f1d95)
- fixed a bug in input file list of haplotypecaller [`7605e7c`](https://github.com/c3g/GenPipes/commit/7605e7cee227bb1a108aed1ea276653cdc4608fa)
- fixed a bug in input file list of haplotypecaller [`3699643`](https://github.com/c3g/GenPipes/commit/36996431f319ef48f2bcbbe47e5a8ccafeb85ac2)
- fixed a bug in input file list of haplotypecaller [`4c686e8`](https://github.com/c3g/GenPipes/commit/4c686e83ea9844d12a1f92cb2ed973ad41080fa5)

## [4.1.2](https://github.com/c3g/GenPipes/compare/4.1.1...4.1.2) - 2022-02-17

### Merged

- remove cluster_submit_cmd from slurm and pbs ini [`#313`](https://github.com/c3g/GenPipes/pull/313)
- Version bump to 4.1.2 [`#312`](https://github.com/c3g/GenPipes/pull/312)
- Release 4.1.2 [`#311`](https://github.com/c3g/GenPipes/pull/311)
- Patch 4.1.2 [`#309`](https://github.com/c3g/GenPipes/pull/309)
- Adjustment to the `prepare_report` step resources in the base ini file, to try to prevent this job getting stuck in the queue. [`#308`](https://github.com/c3g/GenPipes/pull/308)
- Version bump to 4.1.1 [`#306`](https://github.com/c3g/GenPipes/pull/306)

### Commits

- remode empty quotes [`f45fdba`](https://github.com/c3g/GenPipes/commit/f45fdba410e90451e9ec4c8bfaa5cc22a41faf20)
- cleanup queue [`a955e74`](https://github.com/c3g/GenPipes/commit/a955e743ad1fe45f0d145349fea09de6c4268861)
- Updating pipeline READMEs before release [`dd9f185`](https://github.com/c3g/GenPipes/commit/dd9f1853fb2ecad3c18516ce5585da08292c775b)
- PBS gets nodes and ppn together [`4417552`](https://github.com/c3g/GenPipes/commit/4417552683af28edea8faecbca9b4e8e48b9a476)
- GenPipes - Scheduler : Correcting cluster_cpu parsing for PBS [`7dd399f`](https://github.com/c3g/GenPipes/commit/7dd399ffecefd2ce8db586093d75f6597601190d)
- fix hic ini [`df8c9ba`](https://github.com/c3g/GenPipes/commit/df8c9ba5c56bef4c89d28a7e7fdc9627cec13b6e)

## [4.1.1](https://github.com/c3g/GenPipes/compare/4.1.0...4.1.1) - 2022-02-10

### Merged

- Release 4.1.1 [`#305`](https://github.com/c3g/GenPipes/pull/305)
- README.md edited online with Bitbucket [`#304`](https://github.com/c3g/GenPipes/pull/304)
- Version bump to Release 4.1.0 [`#303`](https://github.com/c3g/GenPipes/pull/303)

### Commits

- GenPipes - prep for minor release [`be59560`](https://github.com/c3g/GenPipes/commit/be595604992b4adfbdcac1d2d779c5f8b02dc936)
- Updating READMEs before release [`bbc6156`](https://github.com/c3g/GenPipes/commit/bbc6156de410093af8732e1b83523047295fb51f)
- GenPipes - Tumor Pair : Replacing "xrange" calls by "range" calls to resolve Python3-compatibility issues [`c756a50`](https://github.com/c3g/GenPipes/commit/c756a500f7c587709156057a0709184cc8cd10c8)
- corrected typo in README.md for Nanopore_covseq [`ea152c3`](https://github.com/c3g/GenPipes/commit/ea152c3e2a728af46fe6e27d6b97778ae9142324)
- minor README uedit, again... [`d2a5407`](https://github.com/c3g/GenPipes/commit/d2a5407b152bfd2f927750451634c6314625ae9b)
- Another minor correction in the README... [`21ee999`](https://github.com/c3g/GenPipes/commit/21ee999d77c08529a58cab191538b98f5788f79f)
- Adding missing space in container help message [`77e0698`](https://github.com/c3g/GenPipes/commit/77e0698d8a22797cc01ca571298d4e9061c5b3e0)

## [4.1.0](https://github.com/c3g/GenPipes/compare/4.0.0...4.1.0) - 2022-02-07

### Merged

- Release 4.1 [`#302`](https://github.com/c3g/GenPipes/pull/302)
- Keeping Covseq nanopore up to date for release with latest cleaning and cit ini improvement [`#300`](https://github.com/c3g/GenPipes/pull/300)
- Covseq nanopore latest fixes [`#299`](https://github.com/c3g/GenPipes/pull/299)
- make cluster_cpu scheduler agnostic [`#294`](https://github.com/c3g/GenPipes/pull/294)
- GenPipes - Nanopore : fixing minimap2 and dependencies + allowing duplicate samples (if no duplicate readset) in the readset file [`#298`](https://github.com/c3g/GenPipes/pull/298)
- GenPipes - ChIP-Seq : bash.ini edited with updated versions of software, fixing Issue #127 [`#292`](https://github.com/c3g/GenPipes/pull/292)
- Make walltime and memory ini value more generic [`#289`](https://github.com/c3g/GenPipes/pull/289)
- rename monitor.sh to watchdog [`#290`](https://github.com/c3g/GenPipes/pull/290)
- Covseq ARTIC v4 and v4.1 functionality [`#287`](https://github.com/c3g/GenPipes/pull/287)
- Chipseq atac diffbind [`#288`](https://github.com/c3g/GenPipes/pull/288)
- Version bump to 4.0.0 [`#286`](https://github.com/c3g/GenPipes/pull/286)

### Commits

- Full squash of covseq_ont branch [`e0c9f5f`](https://github.com/c3g/GenPipes/commit/e0c9f5f7859eb69e758bbde0a555f08dcad7531c)
- add cluster_mem to ini option [`ddc1401`](https://github.com/c3g/GenPipes/commit/ddc14010f8ee40dccedeceefccbffb6520f639ac)
- fix regressions [`9d722c2`](https://github.com/c3g/GenPipes/commit/9d722c2596ebaff8b04bcb1e15aa2a801d04c052)
- GenPipes - Release : updating READMEs and VERSION [`ebc18e3`](https://github.com/c3g/GenPipes/commit/ebc18e3f9db3e06336f8ffcfe197ba9e7d0a5c16)
- GenPipes - Release : creation of the release branch [`d00f24a`](https://github.com/c3g/GenPipes/commit/d00f24a31403cce053fb4011adc059c41e38c979)
- covseq_nanopore - Cit - Code cleaning and cit.ini addition [`eaa85d8`](https://github.com/c3g/GenPipes/commit/eaa85d8efe429d7250966e383d2db7e215a54a97)
- methylseq methylkit min cpg issue fix [`dc0680a`](https://github.com/c3g/GenPipes/commit/dc0680a25ac195d00934ac4313257d043b8d4960)
- add walltime format for pbs [`88abd41`](https://github.com/c3g/GenPipes/commit/88abd413d9ac217422f58936a7087e2d2d3f9800)
- update utile time delta [`2f679b7`](https://github.com/c3g/GenPipes/commit/2f679b7b7845a78eac2127473abf901f83b3a700)
- Using cit magic for walltime [`6eddc14`](https://github.com/c3g/GenPipes/commit/6eddc14a4230b5065551ac3c8f9082fa6759a767)

## [4.0.0](https://github.com/c3g/GenPipes/compare/3.6.2...4.0.0) - 2021-12-09

### Merged

- GenPipes - README : update epiqc REAMDE [`#285`](https://github.com/c3g/GenPipes/pull/285)
- Release 4.0 [`#284`](https://github.com/c3g/GenPipes/pull/284)
- GenPipes - Tumor Pair : fixing sym_links steps to avoid duplicate job names [`#282`](https://github.com/c3g/GenPipes/pull/282)
- bug fix #121. missing module load mugqic_tools 2.8.2 [`#281`](https://github.com/c3g/GenPipes/pull/281)
- GenPipes - DNASeq : fixing .bai dependencies [`#280`](https://github.com/c3g/GenPipes/pull/280)
- fixed file path issue in seq2fun and mugqic_tools version in epiQC [`#279`](https://github.com/c3g/GenPipes/pull/279)
- added the reason to use os.remove in a comment [`#278`](https://github.com/c3g/GenPipes/pull/278)
- Epiqc [`#271`](https://github.com/c3g/GenPipes/pull/271)
- Seq2fun denovo rnaseq [`#272`](https://github.com/c3g/GenPipes/pull/272)
- Fixes #96 all GenPipes workflows are now coded as mermaid flowcharts and png generated from them [`#247`](https://github.com/c3g/GenPipes/pull/247)
- Dev covseq documentation update [`#258`](https://github.com/c3g/GenPipes/pull/258)
- log_report.py switching back to py3 and adding Narval as remote [`#277`](https://github.com/c3g/GenPipes/pull/277)
- Hotfix for  DNAseq mpilup - remove recalibrated bam name error from some input jobs [`#276`](https://github.com/c3g/GenPipes/pull/276)
- fixed Issue #118: hicseq hic error in interaction_matrices_Chr (mugqic/genpipes) [`#275`](https://github.com/c3g/GenPipes/pull/275)
- Fix example server name in dev [`#274`](https://github.com/c3g/GenPipes/pull/274)
- Genpipes python3 eh [`#270`](https://github.com/c3g/GenPipes/pull/270)
- Version bump to 3.6.2 [`#269`](https://github.com/c3g/GenPipes/pull/269)
- Fixes #102 issues in README.md for epiqc pipeline [`#253`](https://github.com/c3g/GenPipes/pull/253)

### Fixed

- Merged in seq2fun_fix_bug121 (pull request #281) [`#121`](https://github.com/c3g/GenPipes/issues/121)
- bug fix #121. missing module load mugqic_tools 2.8.2 [`#121`](https://github.com/c3g/GenPipes/issues/121)
- Merged in ss_mermaid_96 (pull request #247) [`#96`](https://github.com/c3g/GenPipes/issues/96)
- Fixes #101 epiQC pipeline workflow added - both manual as well as mermaid generated [`#101`](https://github.com/c3g/GenPipes/issues/101)
- Fixes #102 issues in README.md for epiqc pipeline [`#102`](https://github.com/c3g/GenPipes/issues/102)
- Merged in epiqc_ss (pull request #253) [`#102`](https://github.com/c3g/GenPipes/issues/102)
- Fixes #103 covseq workflow added [`#103`](https://github.com/c3g/GenPipes/issues/103)
- Fixes #101 epiQC pipeline workflow added - both manual as well as mermaid generated [`#101`](https://github.com/c3g/GenPipes/issues/101)
- Fixes #102 issues in README.md for epiqc pipeline [`#102`](https://github.com/c3g/GenPipes/issues/102)
- Fixes #96 all GenPipes workflows are now coded as mermaid flowcharts and png generated from them [`#96`](https://github.com/c3g/GenPipes/issues/96)

### Commits

- addressed Paul's comments [`0c2d569`](https://github.com/c3g/GenPipes/commit/0c2d5693073932083e9720d18be19ea02d2ae254)
- fixed bugs when transferring to python3 [`06e458e`](https://github.com/c3g/GenPipes/commit/06e458ef5f19ceb39d0e68a5da8119db89f129f4)
- fixed fastq concatenate issue with adding zcat if the file is gz [`bdd622d`](https://github.com/c3g/GenPipes/commit/bdd622d0eaf22fa971379afb8aac7c778cd8b8c6)
- added seq2fun protocol [`54bf3aa`](https://github.com/c3g/GenPipes/commit/54bf3aa8ab8425343047f65a4c973af7b2e68d21)
- started pathway analysis-seq2fun [`de5f87b`](https://github.com/c3g/GenPipes/commit/de5f87b3200d1b8fd5adc502dd352ce440ecad6c)
- seq2fun pathway_analysis completed [`dc4c608`](https://github.com/c3g/GenPipes/commit/dc4c608800368464bf3a530469e353fb1f090de4)
- GenPipes - README : Updated READMEs before new release [`8ac6847`](https://github.com/c3g/GenPipes/commit/8ac684765d1d5247c30fe6a597b0f07ef33cae66)
- modified documentation. [`22d90bf`](https://github.com/c3g/GenPipes/commit/22d90bf786b9a706137ea0d5b89c911e794bbd7c)
- updated readme file [`1569435`](https://github.com/c3g/GenPipes/commit/1569435d2afe95b26b82f1228ea20ff6d531aedd)
- updated readme file [`0d1622c`](https://github.com/c3g/GenPipes/commit/0d1622c574e54888af6e34ae38bfb21c483be850)

## [3.6.2](https://github.com/c3g/GenPipes/compare/3.6.1...3.6.2) - 2021-11-11

### Merged

- Release 3.6.2 [`#267`](https://github.com/c3g/GenPipes/pull/267)
- ampliconseq.base.ini edited online with Bitbucket [`#266`](https://github.com/c3g/GenPipes/pull/266)
- Wget warn on fail [`#265`](https://github.com/c3g/GenPipes/pull/265)
- Rnaseq protocol switch fix [`#263`](https://github.com/c3g/GenPipes/pull/263)
- Version bump to 3.6.1 [`#262`](https://github.com/c3g/GenPipes/pull/262)

### Commits

- GenPipes - README : updating READMEs and install scripts before release [`55074b1`](https://github.com/c3g/GenPipes/commit/55074b1d607954fe4e2b17c9c2bf8c44782b51a7)
- GenPipes - Resources : adding some new genome and software install scripts [`1c2d322`](https://github.com/c3g/GenPipes/commit/1c2d32237b906778a1078fd0f113bd56a6643f5d)
- GenPipes - Readset : improved readset parser so that it creates a temptative readset file with unique IDs when the readset file provided has duplicate readert IDs [`c1aef1d`](https://github.com/c3g/GenPipes/commit/c1aef1d93cc61a2e72dbb855a98190776ae93d74)
- GenPipes - MethylSeq : correted bed2interval_list calls + some minor code reformating [`cead55d`](https://github.com/c3g/GenPipes/commit/cead55d8fa0bb1ecec9c11333fc961395627ca16)
- GenPipes - updating version of mugqic_tools in all the pipelines config files [`448d7c8`](https://github.com/c3g/GenPipes/commit/448d7c85a749210355a13f0c0149acf40a847aaf)
- update all pipelines to pandoc 2.16.1 [`cc3f279`](https://github.com/c3g/GenPipes/commit/cc3f279da6e733ea335a46f4cb1f25958cbf60fa)
- new pandoc [`cd9c041`](https://github.com/c3g/GenPipes/commit/cd9c041f094819eb3100dbd254430e5bb7bca2b4)
- Fix protocol mixup for rnaseq pipeline [`2cbdcee`](https://github.com/c3g/GenPipes/commit/2cbdcee19521d60b426d0cac994fb798642824a9)
- Warn when wget command fails. [`beb177e`](https://github.com/c3g/GenPipes/commit/beb177eeb91893b23ebea2a003f780877c36f5ea)
- GenPipes - Readset : correcting the parsing of readset files to stop allowing duplicates headsets in file - fixes Issue #113 [`2fb1dc1`](https://github.com/c3g/GenPipes/commit/2fb1dc1f16595bedf1c788faf825c02c66a8eda1)

## [3.6.1](https://github.com/c3g/GenPipes/compare/3.6.0...3.6.1) - 2021-09-29

### Merged

- Release 3.6.1 [`#261`](https://github.com/c3g/GenPipes/pull/261)
- Fixing sambamba sort output files: there were a bai set as output that was blocking the "resume" for covseq pipeline. [`#260`](https://github.com/c3g/GenPipes/pull/260)
- Version bump to 3.6.0 [`#259`](https://github.com/c3g/GenPipes/pull/259)

### Commits

- Re-creating READMEs before release [`96ce81f`](https://github.com/c3g/GenPipes/commit/96ce81fbb2078db8c3dd9cd84c087aa65812e83e)
- GenPipes - Tumor Pair : fixing issue #106, no more job name overwriting at gym_link_fastq_pair step [`d4613be`](https://github.com/c3g/GenPipes/commit/d4613be2a0f6519bb5cc90bc83083db28cd534fb)
- Updating GenPipes README before release [`0fae24c`](https://github.com/c3g/GenPipes/commit/0fae24c1748830dc11008d1b3818b5b3d8032115)
- more robust log_report [`a026207`](https://github.com/c3g/GenPipes/commit/a026207151232b576827e1ec9daaed30859c6481)
- GenPipes - DNASeq : fixed bed2interval_list call in gatk_haplotype_caller step [`3079e60`](https://github.com/c3g/GenPipes/commit/3079e608f99c0e8770b01a05c92a039aa8805782)
- correct log_report.py before release [`27ccb3b`](https://github.com/c3g/GenPipes/commit/27ccb3b1ee6db0087ab809c495f80f0c44b7efab)
- Merged issue_105_fix_eh into dev [`b472e72`](https://github.com/c3g/GenPipes/commit/b472e728de42402564975f80936d6363f989f510)
- GenPipes - Tumo Pair: correcting job name in sym_link_fastq_pair [`7cd2f09`](https://github.com/c3g/GenPipes/commit/7cd2f0943cd3fe97b3cf49f440fa5eca0391609a)
- GenPipes - Tumor Pair.py : fixing typo in sym_link_fastq_pair [`4df1042`](https://github.com/c3g/GenPipes/commit/4df10424688de33c8f9c1b09c2ae8af502bc6fed)
- remove echo debbug in get wrapper [`b941202`](https://github.com/c3g/GenPipes/commit/b941202d488cc2c0b236734921fad2437c66a2f9)

## [3.6.0](https://github.com/c3g/GenPipes/compare/3.5.0...3.6.0) - 2021-08-30

### Merged

- Release 3.6 [`#250`](https://github.com/c3g/GenPipes/pull/250)
- patch stolen from Ed to fix recalibration step [`#257`](https://github.com/c3g/GenPipes/pull/257)
- Monitor limit [`#256`](https://github.com/c3g/GenPipes/pull/256)
- fix kallisto dependency [`#254`](https://github.com/c3g/GenPipes/pull/254)
- collect metrics fix - dnaseq [`#252`](https://github.com/c3g/GenPipes/pull/252)
- Merge rebasing_tp to dev [`#246`](https://github.com/c3g/GenPipes/pull/246)
- GenPipes - Pipelines : Fixed dnaseq_high_coverage inheritance with dnaseq and rnaseq_light & rnaseq_denovo_assembly inheritance with rnaseq [`#249`](https://github.com/c3g/GenPipes/pull/249)
- Reporting steps addition [`#248`](https://github.com/c3g/GenPipes/pull/248)
- chipseq readme fix on design files [`#245`](https://github.com/c3g/GenPipes/pull/245)
- Updates workflow for chipseq -t chipseq case with differential binding step and dependency update Fixes #93 [`#244`](https://github.com/c3g/GenPipes/pull/244)
- GenPipes - Release 3.5.0 : updating CHANGELOG up to tag 3.5.0 [`#243`](https://github.com/c3g/GenPipes/pull/243)

### Fixed

- Merged in ss_wf_chipseq_93 (pull request #244) [`#93`](https://github.com/c3g/GenPipes/issues/93)
- Updates workflow for chipseq -t chipseq case with differential binding step and dependency update Fixes #93 [`#93`](https://github.com/c3g/GenPipes/issues/93)

### Commits

- GenPipes - Release 3.6 : updating the READMEs in dev [`3e29063`](https://github.com/c3g/GenPipes/commit/3e29063be532b1bff457a37c0b9923ed74424892)
- GenPipes - Tumor Pair : no more attemps to write where the reference files are (in case of readonly FS, e.g. CVMFS...) [`79bc0fd`](https://github.com/c3g/GenPipes/commit/79bc0fdccbb449d5bba2540fc7e20e4599bdc448)
- GenPipes - Tumor Pair : fixed coverage_bed in strelka steps [`ed205e1`](https://github.com/c3g/GenPipes/commit/ed205e15f8919291527916c749cf9cc4fbc2d62a)
- update tp ini for cedar [`203ea90`](https://github.com/c3g/GenPipes/commit/203ea908d6d8db29a82feaefb79c20194abca7d7)
- dev and hs37d5 fixes [`466eb23`](https://github.com/c3g/GenPipes/commit/466eb237183e3997adfcfc1ab0653dd0b5f32acd)
- GenPipes - Tumor Pair : fixed symlink_fastq_pair command [`b6fd739`](https://github.com/c3g/GenPipes/commit/b6fd7399bc764ead51add149450a63b3841fc99a)
- GenPipes - Tumor Pair ; deleted dev.ini from dev branch [`ae4cc08`](https://github.com/c3g/GenPipes/commit/ae4cc084067e621b5b4218888ffb3c8eadf5aeaf)
- removal of dev files for CVMFS version [`7b697cc`](https://github.com/c3g/GenPipes/commit/7b697cc34b91caafdc420a9b1317f39368524e19)
- Rnaseq_light.py: kallisto_count_matrix dependency [`92e4919`](https://github.com/c3g/GenPipes/commit/92e4919440ef7ea651dabaf652e70c84e3960f51)
- cit fixes after dev.ini removal [`f182779`](https://github.com/c3g/GenPipes/commit/f1827795afc91c4932f9ff01e93c9e0442c6071a)

## [3.5.0](https://github.com/c3g/GenPipes/compare/3.4.0...3.5.0) - 2021-07-12

### Merged

- Version bump to 3.5.0 [`#242`](https://github.com/c3g/GenPipes/pull/242)
- Release 3.5 [`#241`](https://github.com/c3g/GenPipes/pull/241)
- Chipseq differential analysis [`#240`](https://github.com/c3g/GenPipes/pull/240)
- Update container [`#238`](https://github.com/c3g/GenPipes/pull/238)
- Fixing hic issue and Upgrading samtools [`#239`](https://github.com/c3g/GenPipes/pull/239)
- Slurm cgroup problem [`#233`](https://github.com/c3g/GenPipes/pull/233)
- Fixing indentation error [`#237`](https://github.com/c3g/GenPipes/pull/237)
- Reporting for covseq [`#235`](https://github.com/c3g/GenPipes/pull/235)
- GenPipes - BFX : updated bash_cmd.py [`#236`](https://github.com/c3g/GenPipes/pull/236)
- Chipseq Minor Fix [`#234`](https://github.com/c3g/GenPipes/pull/234)
- excluding wget call from monitoring loop [`#232`](https://github.com/c3g/GenPipes/pull/232)
- Fixes #83 rnaseq denovo workflow diagram updated for v3.4.0 [`#229`](https://github.com/c3g/GenPipes/pull/229)
- Fixes #82 rnaseq_light workflow updated for v3.4.0 [`#228`](https://github.com/c3g/GenPipes/pull/228)
- Fixes #81 rnaseq workflows are now current to 3.4.0 [`#227`](https://github.com/c3g/GenPipes/pull/227)
- Fixes #80 methylseq pipeline workflow is now current with v3.4.0 [`#226`](https://github.com/c3g/GenPipes/pull/226)
- Fixes #79 hicseq pipeline update for v3.4.0 [`#225`](https://github.com/c3g/GenPipes/pull/225)
- Fixes #78 dnaseq workflow for -t mpileup updated v3.4.0 [`#224`](https://github.com/c3g/GenPipes/pull/224)
- Fixes #77 dnaseq pipeline -t mugqic workflow update for genpipes v3.4.0 [`#223`](https://github.com/c3g/GenPipes/pull/223)
- Fixes #76 dnaseq_highcov pipeline workflow updated v3.4.0 [`#222`](https://github.com/c3g/GenPipes/pull/222)
- Fixes #75 updated amplicon sequence qiime and dada2 workflows for v3.4.0 [`#221`](https://github.com/c3g/GenPipes/pull/221)
- Version bump to 3.4.0 [`#220`](https://github.com/c3g/GenPipes/pull/220)

### Fixed

- Merged in ss_wf_83 (pull request #229) [`#83`](https://github.com/c3g/GenPipes/issues/83)
- Merged in ss_wf_82 (pull request #228) [`#82`](https://github.com/c3g/GenPipes/issues/82)
- Merged in ss_wf_81 (pull request #227) [`#81`](https://github.com/c3g/GenPipes/issues/81)
- Merged in ss_wf_80 (pull request #226) [`#80`](https://github.com/c3g/GenPipes/issues/80)
- Merged in ss_wf_79 (pull request #225) [`#79`](https://github.com/c3g/GenPipes/issues/79)
- Merged in ss_wf_78 (pull request #224) [`#78`](https://github.com/c3g/GenPipes/issues/78)
- Merged in ss_wf_77 (pull request #223) [`#77`](https://github.com/c3g/GenPipes/issues/77)
- Merged in ss_wflow_76 (pull request #222) [`#76`](https://github.com/c3g/GenPipes/issues/76)
- Merged in ss_wflow_75 (pull request #221) [`#75`](https://github.com/c3g/GenPipes/issues/75)
- Fixes #83 rnaseq denovo workflow diagram updated for v3.4.0 [`#83`](https://github.com/c3g/GenPipes/issues/83)
- Fixes #82 rnaseq_light workflow updated for v3.4.0 [`#82`](https://github.com/c3g/GenPipes/issues/82)
- Fixes #81 rnaseq workflows are now current to 3.4.0 [`#81`](https://github.com/c3g/GenPipes/issues/81)
- Fixes #80 methylseq pipeline workflow is now current with v3.4.0 [`#80`](https://github.com/c3g/GenPipes/issues/80)
- Fixes #79 hicseq pipeline update for v3.4.0 [`#79`](https://github.com/c3g/GenPipes/issues/79)
- Fixes #78 dnaseq workflow for -t mpileup updated v3.4.0 [`#78`](https://github.com/c3g/GenPipes/issues/78)
- Fixes #77 dnaseq pipeline -t mugqic workflow update for genpipes v3.4.0 [`#77`](https://github.com/c3g/GenPipes/issues/77)
- Fixes #76 dnaseq_highcov pipeline workflow updated v3.4.0 [`#76`](https://github.com/c3g/GenPipes/issues/76)
- Fixes #76 dnaseq_higcov pipeline workflow updated in sync gpv3.4.0 [`#76`](https://github.com/c3g/GenPipes/issues/76)
- Fixes #75 updated amplicon sequence qiime and dada2 workflows for v3.4.0 [`#75`](https://github.com/c3g/GenPipes/issues/75)
- Fixes #67 Refs #54 dnaseq -t light workflow schema added [`#67`](https://github.com/c3g/GenPipes/issues/67)
- Fixes #65 Refs #54 chipseq pipeline updated [`#65`](https://github.com/c3g/GenPipes/issues/65)
- Fixes #68 Refs #54 nanopore pipeline workflow schema added [`#68`](https://github.com/c3g/GenPipes/issues/68)
- Fixes #66 Dnaseq pipeline -t sv option workflow schema created [`#66`](https://github.com/c3g/GenPipes/issues/66)
- Fixes #60 ampliconseq -t dada2 workflow schema diagram created [`#60`](https://github.com/c3g/GenPipes/issues/60)
- Fixes #53 covseq.py workflow diagram added [`#53`](https://github.com/c3g/GenPipes/issues/53)
- Fixes #51 update covseq pipeline readme to v3.3.0 [`#51`](https://github.com/c3g/GenPipes/issues/51)

### Commits

- started the development. [`3055bd9`](https://github.com/c3g/GenPipes/commit/3055bd9ea8f77fdf799389ec601c201066d73f21)
- started the development. [`4c3b940`](https://github.com/c3g/GenPipes/commit/4c3b9401794469d6cad8f3a54d3f59a9e807907b)
- started the development. [`4e22bde`](https://github.com/c3g/GenPipes/commit/4e22bde84667fdd8cb65ff4af435f5282c822cff)
- update form -n to -c [`a087236`](https://github.com/c3g/GenPipes/commit/a087236b93970b15ca9ada6ab11a1d1de5d57aed)
- update form -n to -c [`1d3bf91`](https://github.com/c3g/GenPipes/commit/1d3bf91794f180ca23043cbb3b58e1103e7af01c)
- Removing guillimin and mammouth inis [`4d85dff`](https://github.com/c3g/GenPipes/commit/4d85dff49edae12ff2f8c81feb3d9851dbafc7b7)
- started the development. [`06b9ddf`](https://github.com/c3g/GenPipes/commit/06b9ddf5cecb1cd565c80c1ee77fc7ea1c8f17aa)
- started the development. [`ad27f00`](https://github.com/c3g/GenPipes/commit/ad27f00ba57bf5b2e4e8abf6a6a315e468a47c49)
- fixed results files creating at custom output folder [`580338d`](https://github.com/c3g/GenPipes/commit/580338df66de511afe783ad2faf01ab9b9a1ad66)
- fixed results files creating at custom output folder [`044565e`](https://github.com/c3g/GenPipes/commit/044565e5381b622266f4e62a410e52add5b2d4be)

## [3.4.0](https://github.com/c3g/GenPipes/compare/3.3.0...3.4.0) - 2021-04-29

### Merged

- Release 3.4 [`#219`](https://github.com/c3g/GenPipes/pull/219)
- Fixes #69 Refs #54 rnaseq -t stringtie workflow schema added [`#214`](https://github.com/c3g/GenPipes/pull/214)
- Refs #67 cleanup after color update [`#216`](https://github.com/c3g/GenPipes/pull/216)
- Ss issu 66 dnaseq sv schema [`#215`](https://github.com/c3g/GenPipes/pull/215)
- Fixes #67 Refs #54 dnaseq -t light workflow schema added [`#212`](https://github.com/c3g/GenPipes/pull/212)
- Update chipseq pipeline schema for -t chipseq and -t atacseq options to reflect latest dev branch pipeline code [`#211`](https://github.com/c3g/GenPipes/pull/211)
- Fixes #68 Refs #54 nanopore pipeline workflow schema added [`#213`](https://github.com/c3g/GenPipes/pull/213)
- Fixes #66 Dnaseq pipeline -t sv option workflow schema created [`#209`](https://github.com/c3g/GenPipes/pull/209)
- Fixing atacseq [`#210`](https://github.com/c3g/GenPipes/pull/210)
- Chipseq design change [`#206`](https://github.com/c3g/GenPipes/pull/206)
- Fix CHICAGO makeDesignFiles call [`#207`](https://github.com/c3g/GenPipes/pull/207)
- VERSION edited online with Bitbucket [`#205`](https://github.com/c3g/GenPipes/pull/205)
- Fixes #60 ampliconseq -t dada2 workflow schema diagram created [`#203`](https://github.com/c3g/GenPipes/pull/203)
- GenPipes - RNAseq : updated star.py to test if the version of STAR is included in the genome_index_folder. If not, use the old format of path (i.e. without the version of STAR) [`#202`](https://github.com/c3g/GenPipes/pull/202)
- GenPipes - RANSeq : updated star.py to add the version of STAR in the genome index folder path [`#200`](https://github.com/c3g/GenPipes/pull/200)
- GenPipes - BFX : corrected delly.py 'call' input handling [Issue 52](https://bitbucket.org/mugqic/genpipes/issues/52/version-330-dnaseq-t-sv) [`#199`](https://github.com/c3g/GenPipes/pull/199)
- covseq pipeline workflow added [`#197`](https://github.com/c3g/GenPipes/pull/197)
- Fixes #51 update covseq pipeline readme to v3.3.0 [`#196`](https://github.com/c3g/GenPipes/pull/196)
- rnaseq.base.ini updated to a newer version of STAR [`#195`](https://github.com/c3g/GenPipes/pull/195)

### Fixed

- Merged in ss_issu_69_rnaseq_strintie_wf (pull request #214) [`#69`](https://github.com/c3g/GenPipes/issues/69)
- Merged in ss_issu_67_wf_dnaseq_light (pull request #212) [`#67`](https://github.com/c3g/GenPipes/issues/67)
- Merged in ss_issu_68_nano_wf (pull request #213) [`#68`](https://github.com/c3g/GenPipes/issues/68)
- Merged in ss_issu_66_dnaseq_sv_schema (pull request #209) [`#66`](https://github.com/c3g/GenPipes/issues/66)
- Fixes #69 Refs #54 rnaseq -t stringtie workflow schema added [`#69`](https://github.com/c3g/GenPipes/issues/69)
- Fixes #65 Refs #54 chipseq pipeline updated [`#65`](https://github.com/c3g/GenPipes/issues/65)
- Fixes #68 Refs #54 nanopore pipeline workflow schema added [`#68`](https://github.com/c3g/GenPipes/issues/68)
- Fixes #67 Refs #54 dnaseq -t light workflow schema added [`#67`](https://github.com/c3g/GenPipes/issues/67)
- Fixes #66 Dnaseq pipeline -t sv option workflow schema created [`#66`](https://github.com/c3g/GenPipes/issues/66)
- Fixes #66 Dnaseq pipeline -t sv option workflow schema created [`#66`](https://github.com/c3g/GenPipes/issues/66)
- Fixes #66 Dnaseq pipeline -t sv option workflow schema created [`#66`](https://github.com/c3g/GenPipes/issues/66)
- Merged in ss_issu_60_wf_ampllconseq (pull request #203) [`#60`](https://github.com/c3g/GenPipes/issues/60)
- Fixes #60 ampliconseq -t dada2 workflow schema diagram created [`#60`](https://github.com/c3g/GenPipes/issues/60)
- Fixes #53 covseq.py workflow diagram added [`#53`](https://github.com/c3g/GenPipes/issues/53)
- Fixes #51 update covseq pipeline readme to v3.3.0 [`#51`](https://github.com/c3g/GenPipes/issues/51)
- Merged in covseq_readme (pull request #196) [`#51`](https://github.com/c3g/GenPipes/issues/51)
- Fixes #51 update covseq pipeline readme to v3.3.0 [`#51`](https://github.com/c3g/GenPipes/issues/51)

### Commits

- GenPipes - Resources : adding software installation scipts [`8c9f613`](https://github.com/c3g/GenPipes/commit/8c9f613904415c3e556c964152a52035a33764a3)
- Updating inis and fixing homer output folder location [`182612e`](https://github.com/c3g/GenPipes/commit/182612e76d8d30d91e99d7efe4298675be8839d9)
- GenPipes - Resources : updates of solftware and genome installation scripts [`d76314e`](https://github.com/c3g/GenPipes/commit/d76314e02de128e3279e41ab58bfff34a09ed5d3)
- Update READMEs [`ef03626`](https://github.com/c3g/GenPipes/commit/ef03626c8f9bef8ccb74d773c77705aa6f142e42)
- GenPipes - Resources : adding Sus_scrofa genome and ivar & kent installation scripts [`e42f490`](https://github.com/c3g/GenPipes/commit/e42f490a6b39d73b449f083e13c26f4f9b3530ae)
- GenPipes - Resources : adding software installation scripts [`ed27f33`](https://github.com/c3g/GenPipes/commit/ed27f331494a7747e0b1ff71406be1d8f8a678cd)
- GenPipes - Release : removing Tumor pair prior release 3.4 [`2b72afb`](https://github.com/c3g/GenPipes/commit/2b72afb3aa39dfb517d3eb5f50d39fa1f675411c)
- more verbose went -d/--design is needed for contrast [`b0875ae`](https://github.com/c3g/GenPipes/commit/b0875ae994a891b481381ef5f090755396bbfadd)
- GenPipes - HiCSeq : corrected typo in CHICAGO makeDesignFiles call [`daf9eb8`](https://github.com/c3g/GenPipes/commit/daf9eb8a64bb6d404e50f6686ada61cff74fc69b)
- force loading mugqic python 2.7 on cuffmerge [`4b08c93`](https://github.com/c3g/GenPipes/commit/4b08c93c3e50ea19291cbe2f01b92d6135c532f8)

## [3.3.0](https://github.com/c3g/GenPipes/compare/3.2.0...3.3.0) - 2021-02-19

### Merged

- Release 3.3 [`#194`](https://github.com/c3g/GenPipes/pull/194)
- Hicseq hicrep pubudu [`#170`](https://github.com/c3g/GenPipes/pull/170)

### Commits

- GenPipes - removing tumor_apir before release [`a4d9a14`](https://github.com/c3g/GenPipes/commit/a4d9a14e2567c03aa0ba9c7b0e586c3693bc2d91)
- Release 3.2 [`ebef4a2`](https://github.com/c3g/GenPipes/commit/ebef4a209be11790e1148d2b98a843bb5cf8f4d7)
- Version bump to 3.2.0 [`63d211d`](https://github.com/c3g/GenPipes/commit/63d211df4e4e8d4532d6f04794ed11f22e21fd2e)
- started to edit the hicseq.py added new file for hicrep [`08dae47`](https://github.com/c3g/GenPipes/commit/08dae47d04452dc7d62be70f988acbc8aa54d49d)
- started to edit the hicseq.py added new file for hicrep [`6775791`](https://github.com/c3g/GenPipes/commit/6775791d3ea6873537ab1d4a7553fb9877926f29)
- started to edit the hicseq.py added new file for hicrep [`6f98cb3`](https://github.com/c3g/GenPipes/commit/6f98cb3a14f5768448916e397adce4435c73ccef)
- started to edit the hicseq.py added new file for hicrep [`6f4136c`](https://github.com/c3g/GenPipes/commit/6f4136c3dab03a02f4aec103a8d56293555c3d9d)
- started to edit the hicseq.py added new file for hicrep [`39b34b0`](https://github.com/c3g/GenPipes/commit/39b34b00ec3780a0a6faa915b8c0d7ed269efe04)
- started to edit the hicseq.py added new file for hicrep [`d664e0e`](https://github.com/c3g/GenPipes/commit/d664e0e5429fbcb2f5504cbe59ab13c960819c6d)
- started to edit the hicseq.py added new file for hicrep [`1c14ddf`](https://github.com/c3g/GenPipes/commit/1c14ddf21996283c3de51dd25176971fee0898f6)

## [3.2.0](https://github.com/c3g/GenPipes/compare/3.1.5...3.2.0) - 2021-01-25

### Merged

- Fixing Issue #46 [`#192`](https://github.com/c3g/GenPipes/pull/192)
- fix: use of undeclared variables [`#187`](https://github.com/c3g/GenPipes/pull/187)
- Fixing rename rename step [`#191`](https://github.com/c3g/GenPipes/pull/191)
- Tumor pair [`#186`](https://github.com/c3g/GenPipes/pull/186)
- tweek monitor and chunk [`#190`](https://github.com/c3g/GenPipes/pull/190)
- Fix out.out in monitor.sh [`#189`](https://github.com/c3g/GenPipes/pull/189)
- PR with latest updates for CovSeq [`#188`](https://github.com/c3g/GenPipes/pull/188)
- Chunk slurm submit [`#185`](https://github.com/c3g/GenPipes/pull/185)
- rnaseq_light.py edited to adjust docstrings to address issue raised by Shaloo here : https://github.com/c3g/GenPipes/issues/63 [`#183`](https://github.com/c3g/GenPipes/pull/183)
- Chunk slurm submit [`#182`](https://github.com/c3g/GenPipes/pull/182)
- replace -a by -r in homer chipseq rsync [`#180`](https://github.com/c3g/GenPipes/pull/180)
- Edited the nanopore ini file to address the mugqic_tools error. [`#179`](https://github.com/c3g/GenPipes/pull/179)
- Added CoVSeQ ini files for Graham and Cedar. Corrected a few errors on the Beluga ini [`#177`](https://github.com/c3g/GenPipes/pull/177)
- remove pacbio from the repo/release [`#176`](https://github.com/c3g/GenPipes/pull/176)
- covseq pipeline [`#175`](https://github.com/c3g/GenPipes/pull/175)
- Nanopore merge [`#173`](https://github.com/c3g/GenPipes/pull/173)
- Check all module in one call [`#164`](https://github.com/c3g/GenPipes/pull/164)
- Added module_perl to the rnaseq_light ini file. [`#171`](https://github.com/c3g/GenPipes/pull/171)
- Add container [`#167`](https://github.com/c3g/GenPipes/pull/167)
- create graham ini file for all pipelines [`#168`](https://github.com/c3g/GenPipes/pull/168)
- Genpipes 3.1.5 release [`#166`](https://github.com/c3g/GenPipes/pull/166)
- watch_portal_folder.py: fix undefined variable [`#165`](https://github.com/c3g/GenPipes/pull/165)
- Carry master bugfixes to dev [`#161`](https://github.com/c3g/GenPipes/pull/161)
- Bug Fix by replacing sacct delimiter | by ^ [`#157`](https://github.com/c3g/GenPipes/pull/157)
- Suggestion, add `module purge` to all jobs that load modules, to avoid conflicts between modules. [`#154`](https://github.com/c3g/GenPipes/pull/154)
- ChIP-Seq - Fix IHEC metrics [`#126`](https://github.com/c3g/GenPipes/pull/126)
- GenPipes - dnaseq.py : bug correction - typo removed [`#142`](https://github.com/c3g/GenPipes/pull/142)
- gatk4.py edited to correct for inconsistencies with configuration parameters within functions. [`#135`](https://github.com/c3g/GenPipes/pull/135)
- Found a bug in the scheduler.py script. I am adding a line to correct it. [`#127`](https://github.com/c3g/GenPipes/pull/127)
- Chipseq atacseq mode + IHEC metrics (merge & report) [`#121`](https://github.com/c3g/GenPipes/pull/121)

### Commits

- Fix line break types [`77126e8`](https://github.com/c3g/GenPipes/commit/77126e8e1b2a64d0e37ba72fdf0fb3dedc4051b6)
- Remove tumor_pair pipeline from release [`1294b96`](https://github.com/c3g/GenPipes/commit/1294b9636a7bf4b2ada2a4f8fa1d23c80ea1a682)
- merging snv and sv, adding protocols [`bd6e6cb`](https://github.com/c3g/GenPipes/commit/bd6e6cb9235f1026263cd0c7c74739b5d6bfd3ba)
- fixes to symlinks for paired indel realignment [`2d3d367`](https://github.com/c3g/GenPipes/commit/2d3d36761a428e6f9b6878fa90919266f54e1ee3)
- cedar ini and exome update [`5475ca4`](https://github.com/c3g/GenPipes/commit/5475ca4a6f1f3667a7cfcc978e1ac4fd6e89db91)
- added multiqc and other tweaks [`e8d90db`](https://github.com/c3g/GenPipes/commit/e8d90db03e2f51475ba1c704d2aa2450f575cb6f)
- dnaseq germline SV updates [`6b52879`](https://github.com/c3g/GenPipes/commit/6b5287904454f782ffe07957330cddc38d82b1ac)
- fixes to sCNAphase [`3731f20`](https://github.com/c3g/GenPipes/commit/3731f205333b6e4505beb39da0144ae1f3744928)
- MGI init commit [`52f8a2b`](https://github.com/c3g/GenPipes/commit/52f8a2b15a85e91135c008724a32db56b0b6db4d)
- sym link dnaseq.base into tumor pair [`6162e8e`](https://github.com/c3g/GenPipes/commit/6162e8e4c9ac5b1942eb4245af8c61814cc41b1b)

## [3.1.5](https://github.com/c3g/GenPipes/compare/3.1.4...3.1.5) - 2020-01-15

### Merged

- GenPipes - BUG correction - trailing space removed in output files of DNASeq skewer trimming step [`#163`](https://github.com/c3g/GenPipes/pull/163)
- BUG correction : corrected bash_cmd.py "ln" function with correct format keys when buiding command [`#162`](https://github.com/c3g/GenPipes/pull/162)
- Bug Correction : corrected 'ln' function in bash_cmd.py to avoid a python "TypeError: cannot concatenate 'str' and 'int' objects" exception [`#158`](https://github.com/c3g/GenPipes/pull/158)
- GenPIpes - DEBUGGING - added slurm-comprehensive walltime for picard_sam_to_fastq in dnaseq.beluga.ini [`#147`](https://github.com/c3g/GenPipes/pull/147)
- GenPipes - DEBUGGING - DNASeq SamToFastq & SymLink steps corrected + working bfx/bash_cmd.py [`#143`](https://github.com/c3g/GenPipes/pull/143)
- GenPIpes - dnaseq.py : bug correction - typo removed [`#141`](https://github.com/c3g/GenPipes/pull/141)
- GenPipes - bug correction in pipeline/dnaseq.py : corrected sym_link_fastq, skewer_trimming & bwa_mem_picard_sort_sam steps, regarding the path of the fastq files when they have to be determined from the readset.bam [`#132`](https://github.com/c3g/GenPipes/pull/132)
- GenPipes - bug correction in pipelines/common.py : corrected the path where the sorted bam files as well as the raw_reads fastq files(from sam_to_fastq) should be written, i.e. within the output directory provided at the pipeline execution [`#131`](https://github.com/c3g/GenPipes/pull/131)
- GenPipes - corrected scheduler.py : removed unwanted sed command in --no-json context [`#129`](https://github.com/c3g/GenPipes/pull/129)
- GenPipes : nanuq2mugqic_pipelines.py : bug corrected - typo in seq_type selection [`#130`](https://github.com/c3g/GenPipes/pull/130)
- Dev [`#125`](https://github.com/c3g/GenPipes/pull/125)
- RNASeq - metrics.py - ihec_metrics_rnaseq : added file in the input_file list to correct job dependency [`#124`](https://github.com/c3g/GenPipes/pull/124)
- Spacesfix [`#117`](https://github.com/c3g/GenPipes/pull/117)
- Add container [`#122`](https://github.com/c3g/GenPipes/pull/122)
- Update to the rnaseq_light README.md file to address issue brought up by GSoD collaborator. [`#120`](https://github.com/c3g/GenPipes/pull/120)
- Cit fix cedar 2019 04 [`#109`](https://github.com/c3g/GenPipes/pull/109)
- README.md edited to remove warning labels. [`#118`](https://github.com/c3g/GenPipes/pull/118)
- Rnaseq rnaseqc Update + verify_bam_id removal [`#116`](https://github.com/c3g/GenPipes/pull/116)
- Eh methylseq single end [`#115`](https://github.com/c3g/GenPipes/pull/115)
- Sanity check mode [`#114`](https://github.com/c3g/GenPipes/pull/114)
- Previous default expression in base.ini would end up basically only retaining transmembrane proteins. Not good. [`#110`](https://github.com/c3g/GenPipes/pull/110)
- Corrected error in the beluga ini for RNAseq de novo [`#103`](https://github.com/c3g/GenPipes/pull/103)
- Corrected minor error in the help message that said that the default protocol was cuflinks. Starting from version 3.1.4, stringtie is the default protocol [`#112`](https://github.com/c3g/GenPipes/pull/112)
- Amplicon pathfix [`#113`](https://github.com/c3g/GenPipes/pull/113)
- Cram [`#111`](https://github.com/c3g/GenPipes/pull/111)
- add check for output job id [`#108`](https://github.com/c3g/GenPipes/pull/108)
- get real path in log_report [`#107`](https://github.com/c3g/GenPipes/pull/107)
- file exist file missing mixup [`#106`](https://github.com/c3g/GenPipes/pull/106)
- bug fixes [`#105`](https://github.com/c3g/GenPipes/pull/105)
- get pipeline info in tsv file can also mute stdout [`#104`](https://github.com/c3g/GenPipes/pull/104)
- run dnaseq multiqc on all samples at once [`#89`](https://github.com/c3g/GenPipes/pull/89)
- add indentation [`#77`](https://github.com/c3g/GenPipes/pull/77)

### Commits

- HICSeq pipeline - improving trimmomatic resources for SULRM + added graham config file for hicseq pipeline [`986f2a3`](https://github.com/c3g/GenPipes/commit/986f2a3c517e5c5e02b2d17495c557e7eafedb08)
- Working singularity version tested on graham [`b787ff7`](https://github.com/c3g/GenPipes/commit/b787ff7edf661a6232f458c003989855ff2af28e)
- Working singularity version tested on graham [`a5b2901`](https://github.com/c3g/GenPipes/commit/a5b2901d2e0ae4f0a7f5fb556a4ab1b4c00a7aaf)
- WIP exec line prototype [`2fe8f47`](https://github.com/c3g/GenPipes/commit/2fe8f4756e05b171ea5ddfbe2942846c92a39cc0)
- WIP exec line prototype [`5229f0e`](https://github.com/c3g/GenPipes/commit/5229f0ee674028e36e3ffa67ed65b952ba8d05e5)
- set cit ini file and config [`733f258`](https://github.com/c3g/GenPipes/commit/733f258a380f33ffa32fb3b6dc9bb5d5b4c916c4)
- add runtime to csv output [`7d9c898`](https://github.com/c3g/GenPipes/commit/7d9c898bf96a3cd60deb5964e07a140c78461cf7)
- Add GENPIPES_CIT env var to run in testing mode [`86e84f3`](https://github.com/c3g/GenPipes/commit/86e84f3eb2c31c2318978d87523d68658d4ae7e6)
- second run of pipeline fixing [`2107daa`](https://github.com/c3g/GenPipes/commit/2107daaf9b4aae17e78128c110397211dc6f3a2f)
- change R_Bioconductor version/cleanup code [`7b25d7b`](https://github.com/c3g/GenPipes/commit/7b25d7bcbba4e5b5f83fa46876230fdb5038720d)

## [3.1.4](https://github.com/c3g/GenPipes/compare/3.1.3...3.1.4) - 2019-03-26

### Merged

- Removed DEV from sleuth configuration [`#75`](https://github.com/c3g/GenPipes/pull/75)
- Corrected slurm parameters for stringtie jobs, including stringtie_abund which was causing errors [`#76`](https://github.com/c3g/GenPipes/pull/76)
- adding draft beluga inis [`#74`](https://github.com/c3g/GenPipes/pull/74)
- parse exit status, and only 50 lines after pro/epilogue [`#73`](https://github.com/c3g/GenPipes/pull/73)
- log report for slurm [`#72`](https://github.com/c3g/GenPipes/pull/72)
- New style class [`#63`](https://github.com/c3g/GenPipes/pull/63)
- Rnaseq jhg [`#69`](https://github.com/c3g/GenPipes/pull/69)
- Rnaseq light jhg [`#68`](https://github.com/c3g/GenPipes/pull/68)
- Methylseq [`#67`](https://github.com/c3g/GenPipes/pull/67)
- Added cedar and mp2b ini files for four pipelines: ampliconseq, dnaseq_high_coverage, pacbio_assembly, rnaseq_light [`#64`](https://github.com/c3g/GenPipes/pull/64)
- Updated version of Bioconductor [`#38`](https://github.com/c3g/GenPipes/pull/38)

### Commits

- pipeline.py  : edited the help so that it actually shows that SLURM is the default scheduler used by GenPipes [`3ef3a04`](https://github.com/c3g/GenPipes/commit/3ef3a043916c71ab3ddead9ab894252d6fdb2c4a)
- Merged master into rnaseq_jhg [`de0d01e`](https://github.com/c3g/GenPipes/commit/de0d01eba1cb634bb6c775ebad2b1ca206217200)
- Merged master into rnaseq_light_jhg [`7293ec0`](https://github.com/c3g/GenPipes/commit/7293ec07ea0ae28d4fb95122605cc90a2258b38b)
- MethylSeq - MethylKit DMR - to fit with C3G code standards, getting rid of bfx/methylkit.py and use bfx/tools.py instead to call methylkit.R, which is part of mugqic_tools [`d1ad834`](https://github.com/c3g/GenPipes/commit/d1ad83446c72d4fb4150c9bb51aeab53b7843fc4)
- corrected Octopus installation script [`0aa9f92`](https://github.com/c3g/GenPipes/commit/0aa9f92ce6a51bb9242957c69faf8fb514c923a6)
- Corrected minor issues with rnaseq_light pipeline, and added documentation to the README [`0d7a7d7`](https://github.com/c3g/GenPipes/commit/0d7a7d7fc04246e3ea3c28bf958eab0a18495aeb)
- MethylSeq - added reference to Bismark in markdown config_and_references.md file for a better report document - also updated the metrics table assignement in MethylSeq.ihec_sample_metrics_report.md and bfx/report/MethylSeq.ihec_sample_metrics_targeted_report.md [`cc01f4e`](https://github.com/c3g/GenPipes/commit/cc01f4ee813c1bf9a960d164f9f055435cfdb16b)
- updated pipeline inis file with the latest version of mugqic_tools i.e. v2.2.2 [`1d490a7`](https://github.com/c3g/GenPipes/commit/1d490a7a3c9b86932481bed4b226aa13f6c56034)
- Updated all the GenPipes .base.ini files with the latest verison of mugqic_R_packages i.e. 1.0.6 [`cea5856`](https://github.com/c3g/GenPipes/commit/cea585601ce652745e73a693fd41375de0015379)
- updated svaba.sh installation script with latest version of SvABA : 1.1.0 [`0ddffdd`](https://github.com/c3g/GenPipes/commit/0ddffdd593dfdf576135927aa42cd837f388e5ee)

## [3.1.3](https://github.com/c3g/GenPipes/compare/3.1.2...3.1.3) - 2018-12-18

### Merged

- Updates GATK4 and annotations [`#47`](https://github.com/c3g/GenPipes/pull/47)
- MethylSeq UMI [`#52`](https://github.com/c3g/GenPipes/pull/52)
- dada2 [`#49`](https://github.com/c3g/GenPipes/pull/49)

### Commits

- added RNA-Seq Light section in the README [`6cc4a37`](https://github.com/c3g/GenPipes/commit/6cc4a37a4ef49c301454b887befeec71e36533bf)
- Added mp2b ini and exome improvements [`91fb349`](https://github.com/c3g/GenPipes/commit/91fb34960ba7ddd053aafc1b9fe1a5ea288e6308)
- Software installation scripts : new ones (mainly for Tumor SV pipeline) and updated ones [`1b3f70c`](https://github.com/c3g/GenPipes/commit/1b3f70c5c7379b37fb567349a1192089b264dd8e)
- Cedar resource and gatk4 metric fixes [`5b912aa`](https://github.com/c3g/GenPipes/commit/5b912aa9d3697fc6d0f91bfee21874a87cb66681)
- MethylSeq pipeline - added MethylKit differential analysis steps & cleaned the code [`26d842c`](https://github.com/c3g/GenPipes/commit/26d842c69b3da4d6d2fca10057615efec5fc4542)
- methylseq.py - commented subroutine all_sample_metrics_report as it has been remove from the pipeline (because useless) [`b29b117`](https://github.com/c3g/GenPipes/commit/b29b1171a29bdad662676c9335182dc8b3752c01)
- GATK4 mp2b file added and improvements to alt contig exclusions [`4177822`](https://github.com/c3g/GenPipes/commit/417782216e1249e6f0a173e157b021d792d9a90e)
- added Ruby, LAST & Picky installation scripts [`ad29c15`](https://github.com/c3g/GenPipes/commit/ad29c150383bda38d5d21ec3035336ce825fd7da)
- BFXDEV-578 - updated FasTree installatino script [`7033838`](https://github.com/c3g/GenPipes/commit/703383837ddc680d6800a590f4c7333d5b83b2b6)
- added methylkit.py as a bfx wrapper to call methylKit.R from mugqic_tools [`e25b313`](https://github.com/c3g/GenPipes/commit/e25b313ba538d14df2371e250140e1949676f7b7)

## [3.1.2](https://github.com/c3g/GenPipes/compare/3.1.1...3.1.2) - 2018-11-21

### Merged

- Hicup arima [`#51`](https://github.com/c3g/GenPipes/pull/51)

### Commits

- Version bump to 3.1.1 [`c6b48be`](https://github.com/c3g/GenPipes/commit/c6b48be693a55e2bae49170d0ec618afac49bd25)
- vcfanno fgbio & delly install scripts [`7909a4e`](https://github.com/c3g/GenPipes/commit/7909a4e37bd780e536d38801efcc040bbf6d1eba)
- Updated R_Bioconductor.sh : pointing to system libraries improved [`2bfc307`](https://github.com/c3g/GenPipes/commit/2bfc30777312e0648a9e88ec170b6658c4c0b21c)
- platypus install script [`9fadba9`](https://github.com/c3g/GenPipes/commit/9fadba98d5c2d020c2d09002fadbbda3653862ed)
- freebayes install script [`f4479be`](https://github.com/c3g/GenPipes/commit/f4479be0eb6d1ef15d87073f5bc4688bdd27ea14)
- removed _dev modules from ini files [`f187cdf`](https://github.com/c3g/GenPipes/commit/f187cdf178cddc6e0ff67659d52f62671b592ed5)
- PacBio Assembly pipeline - corrected incompatibility bug with Guillimin [`e80f729`](https://github.com/c3g/GenPipes/commit/e80f7296159f01cad67679ac27eab5a989fd458d)
- Corrected bug in HicSeq pipeline when running in capture mode [`e8ec0ac`](https://github.com/c3g/GenPipes/commit/e8ec0ac4742a6957d560aa4fba2bbcab93287886)
- editing hicseq.py for Arima compatibility [`6d94474`](https://github.com/c3g/GenPipes/commit/6d9447418ccb10104b7250647e842009d350047e)
- tumor_pair.guillimin.ini : removed some more _dev modules [`31d2d00`](https://github.com/c3g/GenPipes/commit/31d2d00e56a5d07a28bfe094524bb12960bcff6f)

## [3.1.1](https://github.com/c3g/GenPipes/compare/3.1.0...3.1.1) - 2018-11-01

### Merged

- jsonator.py: add version number [`#50`](https://github.com/c3g/GenPipes/pull/50)
- Pipeline stats [`#20`](https://github.com/c3g/GenPipes/pull/20)
- Dashboard display launching user [`#43`](https://github.com/c3g/GenPipes/pull/43)
- Revert "Can run on HPC with slurm and  containers (pull request #41)" [`#46`](https://github.com/c3g/GenPipes/pull/46)
- IHEC metrics [`#44`](https://github.com/c3g/GenPipes/pull/44)
- Add unique md5 checksum to logs [`#42`](https://github.com/c3g/GenPipes/pull/42)
- Can run on HPC with slurm and  containers [`#41`](https://github.com/c3g/GenPipes/pull/41)

### Commits

- Updated pipeline README.md files, with workflow diagram pictures embeded [`11cbf4a`](https://github.com/c3g/GenPipes/commit/11cbf4a83c17a25f477f60cb81ad8b27533df1d7)
- updated pipeline READMEs : all the steps are now shown independently of the available pipeline protocols [`f315771`](https://github.com/c3g/GenPipes/commit/f315771dcac4874815ed2384639a444a2a814f29)
- Added ini files tailored for Mp2b based on Cedar ini files for the following pipelines: RNA-seq, RNA-seq de novo, DNA-seq, Hi-C seq, Methylseq, and ChIP-seq. [`39bf760`](https://github.com/c3g/GenPipes/commit/39bf7606c4b5630c7be050f1a1569ade4965505f)
- Version bump to 3.1.0 [`bd721f1`](https://github.com/c3g/GenPipes/commit/bd721f1d8aae9ff64dfaaa526a80f4856adce2be)
- Python CGI script to tranform pipelines stats log file into a JSON document. [`6f716cb`](https://github.com/c3g/GenPipes/commit/6f716cb934cfcc19f42d8bfa757c0fea946baed3)
- Added dev install scripts for delly, lumpy, sv, vcfanno [`a23ba48`](https://github.com/c3g/GenPipes/commit/a23ba480fbde02414280936571cbaf6c7c487b8e)
- watch_portal_folder: implement update by diff [`2fbdca8`](https://github.com/c3g/GenPipes/commit/2fbdca8830b6650c34442988914a59d05d250c11)
- lint pipeline.py [`7d45ce1`](https://github.com/c3g/GenPipes/commit/7d45ce1edb612fd2b631689f049f5d19fa824169)
- Display JSON log statistics into tables and figures on the log VM. [`86c8a72`](https://github.com/c3g/GenPipes/commit/86c8a724c8e713a21a5c26eb3a4e5d32d4f7a480)
- udpated R_Bioconductor.sh with new packages. Also, now integrates the new installation procedures including pathcing the C3G executables (i.e. use of patchelf) [`63a855e`](https://github.com/c3g/GenPipes/commit/63a855eb2b8cb57057136acecc1cd8a6fe8ef556)

## [3.1.0](https://github.com/c3g/GenPipes/compare/3.0.0...3.1.0) - 2018-03-28

### Merged

- IHEC metrics [`#33`](https://github.com/c3g/GenPipes/pull/33)
- Slurm scheduler bug [`#37`](https://github.com/c3g/GenPipes/pull/37)
- Fix list within list bug which breaks chipseq pipeline. [`#35`](https://github.com/c3g/GenPipes/pull/35)
- Cedar [`#34`](https://github.com/c3g/GenPipes/pull/34)
- Portal integration [`#32`](https://github.com/c3g/GenPipes/pull/32)
- IHEC metrics [`#31`](https://github.com/c3g/GenPipes/pull/31)
- Gary pacbio assembly [`#29`](https://github.com/c3g/GenPipes/pull/29)
- IHEC metrics [`#28`](https://github.com/c3g/GenPipes/pull/28)

### Commits

- updated hicseq.py : calling of the newly built bfx libraries and minor indentations changes [`d03abd5`](https://github.com/c3g/GenPipes/commit/d03abd52b8640664bd36cf4da93a091aefa62021)
- updated pipeline READMEs [`ba5099f`](https://github.com/c3g/GenPipes/commit/ba5099f4a3a2f359c8baa05273419f0df11ea8b0)
- added new bfx libraries to be called by hicseq.py pipeline script [`843d657`](https://github.com/c3g/GenPipes/commit/843d6573f87dad1e4c28c2f543eb63f6054e5ade)
- added some new software installation scripts [`e1dd68f`](https://github.com/c3g/GenPipes/commit/e1dd68f7737d9be17abcbdfc81fd327787e1f7f9)
- added new bash tools wrappers to be called by hicseq.py pipeline script + some minor indentation & syntaxe updates [`e654508`](https://github.com/c3g/GenPipes/commit/e6545082deed4031584d6f45ef0827527212c573)
- updated R_Bioconductor.sh with good indentation and new packages installation [`3f60795`](https://github.com/c3g/GenPipes/commit/3f60795b18d478d30b22dbb904e3aa2f5c74261c)
- add watch_portal_folder.py [`629cef6`](https://github.com/c3g/GenPipes/commit/629cef6195c44c6c8a6928366b63c3ad806010cc)
- updated the bash installation script of python : after the installation of the specified version of Python is successfull, the script now also takes care of the installation of all the needed python libraries [`b02491e`](https://github.com/c3g/GenPipes/commit/b02491e0551b6fafb31cf1db17e05d700b27afcb)
- update DNAseq and link to mugqic_dev [`d26fef6`](https://github.com/c3g/GenPipes/commit/d26fef63a1f86fc84ac9857a417aebe38dfa93fa)
- add chipseq ini file for cedar [`8180609`](https://github.com/c3g/GenPipes/commit/8180609888434102017c03cfa9deb992931f122e)

## [3.0.0](https://github.com/c3g/GenPipes/compare/2.3.0...3.0.0) - 2017-12-07

### Merged

- IHEC metrics [`#27`](https://github.com/c3g/GenPipes/pull/27)
- IHEC metrics [`#26`](https://github.com/c3g/GenPipes/pull/26)
- Methylseq [`#23`](https://github.com/c3g/GenPipes/pull/23)
- Hicseq merge for "capture" functionality [`#25`](https://github.com/c3g/GenPipes/pull/25)
- Hicseq merge [`#22`](https://github.com/c3g/GenPipes/pull/22)
- IHEC metrics [`#21`](https://github.com/c3g/GenPipes/pull/21)

### Commits

- Version bump to 3.0.0 [`03638bd`](https://github.com/c3g/GenPipes/commit/03638bd37001471dfa6802bd0e7fd387fa360d45)
- Updated README for all the pipelines [`7e14f09`](https://github.com/c3g/GenPipes/commit/7e14f0918ca4b76c34bdc3545364efb7db9fb966)
- Version bump to 3.0.0 [`03f3dda`](https://github.com/c3g/GenPipes/commit/03f3dda38dcd2bff8bec6ca23655f7f391ff11b4)
- minor updates on bfx/tools.py especially to make indentation uniform across the whole file [`29e74ac`](https://github.com/c3g/GenPipes/commit/29e74ac587d67c6757446679b6ba4094c554527b)
- RNA-Seq pipeline - updated .base.ini file : removed specific module loading for differential expression step + minor indentation updates within trinity.py for RNA-Seq de-novo Assembly pipeline [`00eb96b`](https://github.com/c3g/GenPipes/commit/00eb96b9ac8d5af600b2a3ee6ecd4357b8a23e00)
- homer edits to generalize methods [`ac73327`](https://github.com/c3g/GenPipes/commit/ac73327c82c02fb6abd581cebd3c2b65c2c74a9d)
- commit prior merging to avoid conflict [`3d964b2`](https://github.com/c3g/GenPipes/commit/3d964b2b8f23390e0b37af97253a771f69c83e47)
- changes to mammouth.ini to set all ppn=1; changed module spider back to module show since it is incompatible with abacus [`130d0b5`](https://github.com/c3g/GenPipes/commit/130d0b56c36e89b87918c36a568730ce9255c54c)
- PacBio Assembly Pipeline - updated circlator step : created a circlator.py within the bfx folder and review the coding/calling of the circlator step within the pipeline python wrapper [`bfdcbae`](https://github.com/c3g/GenPipes/commit/bfdcbae60e2085741b699ee92dc27279212a5521)
- BFXDEV-673 - adding analysis JSON file generation to the RNA-Seq pipeline [`fdddddb`](https://github.com/c3g/GenPipes/commit/fdddddb7aa49ec5c727d61b1ddc706fdcdace921)

## [2.3.0](https://github.com/c3g/GenPipes/compare/2.2.1...2.3.0) - 2017-02-27

### Commits

- Numerous speed improvements and addition of fast variant calling [`9729909`](https://github.com/c3g/GenPipes/commit/9729909b73e8b9d3112af94ce86e5d24d4728a97)
- updates brought to many module install scripts, and adding of new modules [`06e095f`](https://github.com/c3g/GenPipes/commit/06e095f694245ecbe018e0d9f7ade055acc427fb)
- Pre-release - adding the installation scripts for all the new software used by tumor_pair pipeline [`7717c35`](https://github.com/c3g/GenPipes/commit/7717c35ebe260f304a0c75f422e218d7f98f33db)

## [2.2.1](https://github.com/c3g/GenPipes/compare/2.2.0...2.2.1) - 2016-12-19

### Merged

- Irp_modifications [`#17`](https://github.com/c3g/GenPipes/pull/17)

### Commits

- PICARD 2.0.1 - added picard2.py to the bfx tools [`916b8d6`](https://github.com/c3g/GenPipes/commit/916b8d6a22377df55432c5587eee87285e1b27e1)
- remove conflict [`7e28490`](https://github.com/c3g/GenPipes/commit/7e2849060822c19714483b0059bb5beb0d612df1)
- new reference genome installation scripts (new species) [`3d0d765`](https://github.com/c3g/GenPipes/commit/3d0d765e22a714293493d33f310771818c8bb72d)
- qiime db installation scripts and some new module scipts [`16f2ae7`](https://github.com/c3g/GenPipes/commit/16f2ae774eb5fbcea22b53634d92c2debff5e680)
- new module installation scripts [`ad09a1e`](https://github.com/c3g/GenPipes/commit/ad09a1ee8b1094f2926837548eb698a8d50e4cf9)
- commits before merging [`6330a7f`](https://github.com/c3g/GenPipes/commit/6330a7f4570a90fe96d3d8360a84c470b67c2285)
- updating module and genome scripts from mammouth [`8ab80b3`](https://github.com/c3g/GenPipes/commit/8ab80b30e8b6192cf4623f3fcebacbed40c733f4)
- some updated & new module installation scripts [`a4f4b05`](https://github.com/c3g/GenPipes/commit/a4f4b053fc056dc3eb001d333aa301907800d354)
- RNASeq de-novo assembly - updated trinity call with newer version of trinity and new default parameter [`8f88daa`](https://github.com/c3g/GenPipes/commit/8f88daae6fe4db7c1c418a8c8ba29f52241b8331)
- RNASeq - allow sample names to have '.' within their name without crashing during gq_seq_utils_exploratory_analysis_rnaseq [`cece203`](https://github.com/c3g/GenPipes/commit/cece203d76ad1bc185f93fadc7805ea6ae62574a)

## [2.2.0](https://github.com/c3g/GenPipes/compare/2.1.1...2.2.0) - 2016-02-08

### Merged

- Irp_end_fastq_notification [`#14`](https://github.com/c3g/GenPipes/pull/14)
- Irp_bcl2fastq2 [`#13`](https://github.com/c3g/GenPipes/pull/13)
- Rnaseq_denovo_assembly_new_features [`#10`](https://github.com/c3g/GenPipes/pull/10)
- Illumina_run_processing_sprint [`#11`](https://github.com/c3g/GenPipes/pull/11)
- Pacbio_whitelist [`#7`](https://github.com/c3g/GenPipes/pull/7)
- ingstats.py - Adding Array callrate check (Missing or Low) and add support for different manifest version through the -m option - BFXDEV-445 - BFXDEV-466 - BFXDEV-451 [`#8`](https://github.com/c3g/GenPipes/pull/8)

### Commits

- Version bump to 2.2.0 [`477b332`](https://github.com/c3g/GenPipes/commit/477b3324622d7dd3f2f9893de073d3a95ad798c5)
- resolving conflict [`db54b51`](https://github.com/c3g/GenPipes/commit/db54b51a452ad056439a98dfb5fd0efe31729e35)
- updated ini files - BFXDEV-490 [`442830d`](https://github.com/c3g/GenPipes/commit/442830db33412c70b263168401491b720c9f1ac6)
- resolving conflict [`3a2c22f`](https://github.com/c3g/GenPipes/commit/3a2c22faf9ce6b495d6e5a4cba90422ae87f729a)
- resolving conflict [`228c23a`](https://github.com/c3g/GenPipes/commit/228c23afb094d29cc50883e760d85d386e520fee)
- DNASEQ - update SnpEff command for versuion 4.2 - BFXDEV-490 [`4103b9f`](https://github.com/c3g/GenPipes/commit/4103b9fb375d9b4fff37619c39c5f405bd488d06)
- DNASEQ - allow filter_nstrech to use non recalibrated compress data && update SnpEff command for versuion 4.2 - BFXDEV-505 ; BFXDEV-490 [`71c233c`](https://github.com/c3g/GenPipes/commit/71c233cc4262bb06589c94c59a1bc99d5ab73ab4)
- DNASEQ - allow filter_nstrech to use non recalibrated compress data && update SnpEff command for versuion 4.2 - BFXDEV-505 ; BFXDEV-490 [`b041650`](https://github.com/c3g/GenPipes/commit/b041650c59a30b8473d19fa91f23326541784134)
- DNASEQ - support single end libray for metrics and dnasample metrics steps & allow filter_nstrech to use non recalibrated compress data - BFXDEV-503 BFXDEV-505 [`8e1bbc2`](https://github.com/c3g/GenPipes/commit/8e1bbc27f699182ff5ae64e519692ebba4b399fc)
- DNASEQ - support single end libray for metrics and dnasample metrics steps - BFXDEV-503 [`ffcee7d`](https://github.com/c3g/GenPipes/commit/ffcee7d8c3988ce50253b295f07b30be4f6a695c)

## [2.1.1](https://github.com/c3g/GenPipes/compare/2.1.0...2.1.1) - 2015-04-13

### Commits

- Added GNU Lesser General Public License (LGPL) to MUGQIC Pipelines [`5e335b4`](https://github.com/c3g/GenPipes/commit/5e335b49128cee62ffec7caa307982820026b2da)
- BFXDEV-59 ChIP-Seq report redesign [`01a41aa`](https://github.com/c3g/GenPipes/commit/01a41aac0280764a35f9f394a8afc4219bf57d45)
- BFXDEV-59 Completed RNA-Seq report redesign [`c183fad`](https://github.com/c3g/GenPipes/commit/c183fadef7615efb856be84214cd44bcffbde5c7)

## [2.1.0](https://github.com/c3g/GenPipes/compare/2.0.2...2.1.0) - 2015-02-04

### Commits

- README.md edited online with Bitbucket [`af39fb4`](https://github.com/c3g/GenPipes/commit/af39fb42a2f0251bb2d27c6160248223c3c531cf)
- BFXDEV-292 First draft of chipseq pipeline [`51f3444`](https://github.com/c3g/GenPipes/commit/51f34444b28fe30eafce3d44245acf8f81decf94)
- BFXDEV-292 Last steps implementation in chipseq pipeline [`5017938`](https://github.com/c3g/GenPipes/commit/5017938f468fa34a24e15a2e4701c7db0a2f4aee)

## [2.0.2](https://github.com/c3g/GenPipes/compare/2.0.1...2.0.2) - 2015-01-12

### Commits

- Standardized MACS2 module install [`ddbd8f2`](https://github.com/c3g/GenPipes/commit/ddbd8f2ee1da7eb2b6b522f03b51ec86afb0bd23)
- Version bump to 2.0.2 [`104778a`](https://github.com/c3g/GenPipes/commit/104778ab333aeb550212f4ac66bb6e2e7f851093)
- Standardized bowtie2 module install [`94910f9`](https://github.com/c3g/GenPipes/commit/94910f99c7d2372b84e6ac2bb3d342b2bd4b91c5)

## [2.0.1](https://github.com/c3g/GenPipes/compare/2.0.0...2.0.1) - 2014-12-17

### Commits

- PacBio Assembly README.md generated by --help [`3a17272`](https://github.com/c3g/GenPipes/commit/3a17272ef0fca309f47ca4b4d42f54911fdb2d72)
- BFXDEV-296 Added DNA-Seq step docstring documentation [`9a53a11`](https://github.com/c3g/GenPipes/commit/9a53a1158ac2177705e22ecaf215b686afef397a)
- BFXDEV-297 Add RNA-SeQC in run processing pipeline [`7e38acf`](https://github.com/c3g/GenPipes/commit/7e38acffb6de319ad419bfb6dce05711ee836838)

## [2.0.0](https://github.com/c3g/GenPipes/compare/1.4...2.0.0) - 2014-12-11

### Commits

- Reorganized all config parameters [`475e92c`](https://github.com/c3g/GenPipes/commit/475e92cb4bf74e4f3fc763d7231276a4aac7d1d2)
- Run processing: Merge md5 jobs (fastq+bam) and run qc graphs on BAM (or fallback to fastq) [`7bb7a50`](https://github.com/c3g/GenPipes/commit/7bb7a5066e26e6be08ba96be21dddd09e884ef34)
- Updated PacBio pipeline steps as docstrings [`6eb4df2`](https://github.com/c3g/GenPipes/commit/6eb4df27c316fd9c5353f43a9962297258b8d6d6)

## [1.4](https://github.com/c3g/GenPipes/compare/1.3...1.4) - 2014-11-17

### Commits

- Indentation correction. BFXDEV-31 [`a996d3f`](https://github.com/c3g/GenPipes/commit/a996d3f19d1bb57ee86aae1b6d80387b4cecffee)
- Updated config genome paths according to new genome organization [`c2d5000`](https://github.com/c3g/GenPipes/commit/c2d50007704aa3de794edd75e4d99b97dad76eb6)
- Updated parameters [`1748e1a`](https://github.com/c3g/GenPipes/commit/1748e1a1e848641c4a7bdf1b84fa44a7984be8dc)

## [1.3](https://github.com/c3g/GenPipes/compare/1.2...1.3) - 2014-06-02

### Commits

- updated ini files. BFXDEV-31 [`7b8a64c`](https://github.com/c3g/GenPipes/commit/7b8a64ce88239db4e945067775222261e73265d5)
- More Trinotate steps in RNASeq De Novo [`0194584`](https://github.com/c3g/GenPipes/commit/0194584531fe8c28ce7bbd179bc703306f118fbe)
- added / updated ini files. [`d79cc38`](https://github.com/c3g/GenPipes/commit/d79cc38859819fc336e527c4eba924d371b28ae6)

## [1.2](https://github.com/c3g/GenPipes/compare/1.1...1.2) - 2014-03-28

### Commits

- updated ini files for RRNATagger. BFXDEV-31 [`c578e83`](https://github.com/c3g/GenPipes/commit/c578e8380112e319fd6ccba5abf97bf97ca15c26)
- rnaSeq.guillimin_hg19.ini [`2172223`](https://github.com/c3g/GenPipes/commit/21722231dacabf3c60cfcd76dfbb87137ee56737)
- rnaSeq.guillimin_mm10.ini [`d75c72d`](https://github.com/c3g/GenPipes/commit/d75c72d74744c34f9b144fd9696797db3d65055f)

## [1.1](https://github.com/c3g/GenPipes/compare/1.0...1.1) - 2013-12-23

### Commits

- Cleanup of rnaseq_denovo files [`e61eb0f`](https://github.com/c3g/GenPipes/commit/e61eb0f302eaef7234971f3e85f1f38f6f2db498)
- Major modifications to the pacbio pipeline. Functional version tested [`bb5a19e`](https://github.com/c3g/GenPipes/commit/bb5a19ea0e484f3c4efcee896086e61021d54775)
- Modifications to perl packages for Pacbiopipeline. BFXDEV-30 [`b430c4c`](https://github.com/c3g/GenPipes/commit/b430c4cead51de26b5e47ad704fa8124fa93228b)

## 1.0 - 2013-11-08

### Commits

- remove tools_shed for the new tool repo [`6db6f70`](https://github.com/c3g/GenPipes/commit/6db6f703507d8566bfc75be2a85416885b4ce633)
- Moved module and genome files to mugqic_resources repository [`06ca9c5`](https://github.com/c3g/GenPipes/commit/06ca9c5df052282be414cccedad5fffe6479f6ba)
- A bit of a cleanup [`32f25ea`](https://github.com/c3g/GenPipes/commit/32f25ea9c63f28988f1221c29b106045fe312560)
