### Differential Expression Analysis - Methods

#### Differential Analysis Design

The experimental design resulted from a discussion with the client. The designs used in differential analysis are presented in the following table, which contains the sample names as well as the sample group membership per design. For each experimental design (column name), three conditions/groups are possible: 0, 1 and 2. If a sample is assigned 0, it is not included in a particular analysis. If a sample is assigned 1, the sample is considered as a member of the control group. If a sample is assigned 2, the sample is considered as a member of the test/case group.

Table: Sample Names and Experimental Designs (**partial table**; [download full table](design.tsv))

$design_table$


#### Differential Gene Analysis Description

A primary task in the analysis of RNA-Seq data is the detection of differentially expressed genes. For this purpose, count data from RNA-Seq should be obtained for each non-overlapping gene. Read counts are found to be (to good approximation) linearly related to the abundance of the target transcript [@quantifying_rnaseq]. If reads were independently sampled from a population with given, fixed fractions of genes, the read counts would follow a multinomial distribution, which can be approximated by the Poisson distribution. Thus, we can use statistical testing to decide whether, for a given gene, an observed difference in read counts is significant, that is, whether it is greater than what would be expected just due to natural random variation. The differential gene expression analysis is done using [DESeq]\ [@deseq] and [edgeR]\ [@edger] R Bioconductor packages

Read counts are obtained using [HTSeq] and are represented as a table which reports, for each sample (columns), the number of reads mapped to a given gene (rows).

Table: Matrix of Raw Read Counts per Gene per Sample (**partial table**; [download full table](rawCountMatrix.csv))

$raw_count_matrix_table$


#### Differential Transcripts Analysis Description

FPKM values calculated by [Cufflinks] are used as input. The transcript quantification engine of Cufflinks, [Cuffdiff]\ [@cufflinks], is used to calculate transcript expression levels in more than one condition and test them for significant differences. To identify a transcript as being differentially expressed, Cuffdiff tests the observed log-fold-change in its expression against the null hypothesis of no change (i.e. the true log-fold-change is zero). Because measurement errors, technical variability, and cross-replicate biological variability might result in an observed log-fold-change that is nonzero, Cuffdiff assesses significance using a model of variability in the log-fold-change under the null hypothesis. This model is described in details in [@cuffdiff].


#### Gene Ontology (GO) Analysis of the Differential Expression Results

One simple, but extremely widely used, systems biology technique for highlighting biological processes is to look at gene category the show over-representation in the differential analysis results. In order to perform this analysis, genes are grouped into categories defined by the common biological properties and then tested to find categories that are over-represented amongst the differentially expressed genes. Gene Ontology (GO) terms are the most commonly used categories. RNA-Seq data suffers from a bias in detecting differential expression for long genes. This means that when using a standard analysis, any category containing a preponderance of long genes will be more likely to show up as being over-represented than a category containing genes of average lengths. Thus the GO analysis is performed using of the [goseq]\ [@goseq] R Bioconductor package which provides methods for performing Gene Ontology analysis of RNA-Seq data, taking this length bias into account.


### Differential Expression Analysis - Results

The following sections provide the results of the differential gene expresssion and differential transcript expression per design.

The results of the differential gene expression analysis are generated using [DESeq] and [edgeR]. Raw read counts generated by [HTSeq] are used as input. The first column is the gene id; the second column is the gene symbol; the third column is the log2 fold change of the level of expression of each gene; the fourth column is the nominal p-value of the test; and the fifth column is the FDR adjusted p-value of the test.

The results of the differential transcript expression analysis are generated using [Cuffdiff]. FPKM values calculated by Cufflinks are used as input. The first column is the transcript id, the second column is the gene symbol, the third column is the log2 fold change of the level of expression of each transcript, the fourth column is the nominal p-value of the test and the fifth column is the FDR adjusted p-value of the test.

The results of the differential gene and transcript expression analyses are processed in a gene ontology analysis using goseq. There are two gene ontology analyses per design. The first one is based on the differentially expressed genes from the DESeq/edgeR results. Every gene having an FDR adjusted p-value under $adj_pvalue_threshold$ is considered as differentially expressed. The second analysis is based on the differentially expressed transcripts from the cuffdiff results. The threshold for differentially expressed selection is the same as in the first analysis. The first column is the ID of the category enriched; the second column is the FDR adjusted p-value of category enrichment; the third column is the GO id; and the fourth column is a brief description of the GO term. **If there are no results for a particular design, it means the FDR adjusted GO enrichment was not significant (p-value too high)**.
