#### Gene Ontology (GO) Analysis of the Differential Expression Results

One simple, but extremely widely used, systems biology technique for highlighting biological processes is to look at gene category the show over-representation in the differential analysis results. In order to perform this analysis, genes are grouped into categories defined by the common biological properties and then tested to find categories that are over-represented amongst the differentially expressed genes. Gene Ontology (GO) terms are the most commonly used categories. RNA-Seq data suffers from a bias in detecting differential expression for long genes. This means that when using a standard analysis, any category containing a preponderance of long genes will be more likely to show up as being over-represented than a category containing genes of average lengths. Thus the GO analysis is performed using the [goseq]\ [@goseq] R Bioconductor package which provides methods for performing Gene Ontology analysis of RNA-Seq data, taking this length bias into account. 

One gene ontology analysis is generated per design based on the differentially expressed genes from the DESeq/edgeR results. GO terms associated to each gene are obtained from trinotate annotations. Every gene having an FDR adjusted p-value under 0.05 is considered as differentially expressed. The threshold for differentially expressed selection is the same as in the first analysis. The first column is the ID of the category enriched; the second column is the FDR adjusted p-value of category enrichment; the third column is the GO id; and the fourth column is a brief description of the GO term. **If there are no results for a particular design, it means the FDR adjusted GO enrichment was not significant (p-value too high)**.