```{r preparation, results="hide",echo=F,message=F,warning=F}

### DEFINED OUTSIDE
# report_dir
# source_dir
# contrasts
# top_n_results
# design_file

library(rmarkdown)
library(knitr)
library(hwriter)
library(magrittr)
options(stringsAsFactors=F)



# Directories and cp data
#unlink( file.path(report_dir,"DiffExp","genes") , recursive = T)
dir.create( file.path(report_dir), showWarnings=F,recursive=T)
# copy  design, dge results, counts
file.copy(design_file, file.path(report_dir), overwrite = T)
#flist <- list.files(source_dir, "^.+counts.matrix.symbol$", full.names = T)
#flist <- c(flist, list.files(source_dir, "ontology_results.csv$", full.names = T, recursive=T) )
#flist <- c(flist, list.files(source_dir, "dge_trinotate_results.csv$", full.names = T, recursive=T) )
#print(paste("NOTICE: Copying the following files", paste(flist, collapse=" , ") ))
#file.copy(flist, report_dir)

file.copy(from = source_dir, to = report_dir, overwrite = T, recursive = T)


design = design_file %>% read.delim(header=T,check.names=F,colClasses="character",stringsAsFactors=F,comment.char='')

# Create dge/ GoSeq results tables
goseq.results=data.frame(NULL)
dge.results=data.frame(NULL)
counts.results=data.frame(NULL)
flist <-c()
for (item in c("genes","isoforms")){
    summary = file.path(source_dir,paste(item,".counts.matrix.symbol",sep="")) %>% read.delim(header=T,check.names=F,colClasses="character",stringsAsFactors=F,comment.char='')
    summary = summary[c(1:min(top_n_results,nrow(summary))),c(1:min(ncol(summary),8))]
    names(summary)[1]="id"
    counts.results = rbind(counts.results, cbind(fileID=item, summary))    
    for (contrast in contrasts){
        # table is sorted by default, take the first 7 lines    
        summary = file.path(report_dir,"differential_expression",item,contrast,"gene_ontology_results.csv") %>% read.delim(header=T,check.names=F,colClasses="character",stringsAsFactors=F,comment.char='')        
        summary = summary[c(1:min(top_n_results,nrow(summary))),c(1:min(ncol(summary),8))] 
        goseq.results=rbind(goseq.results, cbind(fileID=paste(item,contrast,sep="_"), summary))
        summary = file.path(report_dir,"differential_expression",item,contrast,"dge_trinotate_results.csv") %>% read.delim(header=T,check.names=F,colClasses="character",stringsAsFactors=F,comment.char='')
        # table is sorted by default, take the first 7 lines    
        summary = summary[c(1:min(top_n_results,nrow(summary))),c(1:min(ncol(summary),8))]
        dge.results=rbind(dge.results, cbind(fileID=paste(item,contrast,sep="_"), summary))        
    }        
}

```

### Differential Expression Analysis - Methods

#### Differential Analysis Design

The experimental design resulted from a discussion with the client. The designs used in differential analysis are presented in the following table, which contains the sample names as well as the sample group membership per design. For each experimental design (column name), three conditions/groups are possible: 0, 1 and 2. If a sample is assigned 0, it is not included in a particular analysis. If a sample is assigned 1, the sample is considered as a member of the control group. If a sample is assigned 2, the sample is considered as a member of the test/case group.

```{r design_table, results='asis', echo=F,message=F,warning=F}
print(kable(design, row.names=F, caption=paste("<br>","Sample Names and Experimental Designs", 
                                                   " (**partial table**; [download full table](",basename(design_file),"))",sep=""))) 
```            

#### Differential Gene Analysis Description

A primary task in the analysis of RNA-Seq data is the detection of differentially expressed genes. For this purpose, count data from RNA-Seq should be obtained for each non-overlapping gene. Read counts are found to be (to good approximation) linearly related to the abundance of the target transcript [@quantifying_rnaseq]. If reads were independently sampled from a population with given, fixed fractions of genes, the read counts would follow a multinomial distribution, which can be approximated by the Poisson distribution. Thus, we can use statistical testing to decide whether, for a given gene, an observed difference in read counts is significant, that is, whether it is greater than what would be expected just due to natural random variation. The differential gene expression analysis is done using [DESeq]\ [@deseq] and [edgeR]\ [@edger] R Bioconductor packages

Gene abundance estimation is performed using [RSEM]\ [@rsem] and is represented as a table which reports, for each sample (columns), the number of reads mapped to a given gene (rows).

```{r counts_table, results='asis', echo=F,message=F,warning=F}
for (item in c("genes","isoforms")){        
    subset.counts=try(subset(counts.results, fileID == item, select=-fileID))
    if (!inherits(subset.counts,'try-error') && nrow(subset.counts) > 0){            
        print(kable(subset.counts, row.names=F, caption=paste("<br>","Matrix of Raw Read Counts per ",item," per Sample ",  
                                                              " (**partial table**; [download full table](","differential_expression/",item,".counts.matrix.symbol))",sep=""), 
        padding = 10, align=NULL))
        }else  print(paste("<br>","No abundance estimation ",item," results are available .",sep=""))    
}        
    
``` 

#### Gene Ontology (GO) Analysis of the Differential Expression Results

One simple, but extremely widely used, systems biology technique for highlighting biological processes is to look at gene category the show over-representation in the differential analysis results. In order to perform this analysis, genes are grouped into categories defined by the common biological properties and then tested to find categories that are over-represented amongst the differentially expressed genes. Gene Ontology (GO) terms are the most commonly used categories. RNA-Seq data suffers from a bias in detecting differential expression for long genes. This means that when using a standard analysis, any category containing a preponderance of long genes will be more likely to show up as being over-represented than a category containing genes of average lengths. Thus the GO analysis is performed using the [goseq]\ [@goseq] R Bioconductor package which provides methods for performing Gene Ontology analysis of RNA-Seq data, taking this length bias into account. 

One gene ontology analysis is generated per design based on the differentially expressed genes from the DESeq/edgeR results. GO terms associated to each gene are obtained from trinotate annotations. Every gene having an FDR adjusted p-value under 0.05 is considered as differentially expressed. The threshold for differentially expressed selection is the same as in the first analysis. The first column is the ID of the category enriched; the second column is the FDR adjusted p-value of category enrichment; the third column is the GO id; and the fourth column is a brief description of the GO term. **If there are no results for a particular design, it means the FDR adjusted GO enrichment was not significant (p-value too high)**.

#### Differential Expression Analysis and Gene Ontology (GO) Analysis of the Differential Expression - Results

The following sections provide the results of the differential gene expresssion per design. 

* Gene: component id
* Symbol: BLAST protein database id if available, component id otherwise
* log_FC: log2 Fold Change of gene level expression
* log_CPM: log2 Counts Per Million of gene level expression
* deseq.p-value: DESeq nominal p-value
* deseq.adj.pvalue: DESeq False Discovery Rate (FDR) adjusted p-value
* edger.p-value: edgeR nominal p-value
* edger.adj.pvalue: edgeR False Discovery Rate (FDR) adjusted p-value


```{r summary_table, results='asis', echo=F,message=F,warning=F}

for (item in c("genes","isoforms")){    
    for (contrast in contrasts){
        subset.dge=try(subset(dge.results, fileID == paste(item,contrast,sep="_"), select=-fileID))
        if (!inherits(subset.dge,'try-error') && nrow(subset.dge) > 0){            
            print(kable(subset.dge, row.names=F, caption=paste("<br>","Differentially Expressed ",item," for contrast ", contrast, 
                                                               " (**partial table**; [download full table](","differential_expression/", item,"/",contrast,"/dge_trinotate_results.csv))",sep=""), 
                                    padding = 10, align=NULL))
            }else  print(paste("<br>","No differentially expressed ",item," results are available for this contrast after filtering: ", contrast,  ".",sep=""))    
        
        subset.goseq=try(subset(goseq.results,fileID == paste(item,contrast,sep="_"), select=-fileID))
        if (!inherits(subset.goseq,'try-error') && nrow(subset.goseq) > 0){            
            print(kable(subset.goseq, row.names=F, caption=paste("<br>","GO Results of the Differentially Expressed ",item," for contrast ", contrast, 
                                                                 " (**partial table**; [download full table](","differential_expression/", item,"/", contrast,"/gene_ontology_results.csv))",sep=""), 
                        padding = 10, align=NULL))
            } else print(paste("<br>","No FDR adjusted GO enrichment was significant (p-value too high) based on the differentially expressed ",item, " results for this contrast: ", contrast ,".",sep=""))            
    }
}
```