```{r preparation, results="hide",echo=F,message=F,warning=F}

### DEFINED OUTSIDE
# report_dir
# source_dir

library(rmarkdown)
library(knitr)
library(hwriter)
library(magrittr)
options(stringsAsFactors=F)


# Directories and cp data
dir.create( file.path(report_dir), showWarnings=F,recursive=T)
zip( file.path(report_dir,"trinotate_annotation_report.zip"), files = file.path(source_dir,"trinotate_annotation_report.tsv") )

```

### Trinotate Annotation

Trinotate is a comprehensive annotation suite designed for automatic functional annotation of transcriptomes. Trinotate makes use of a number of different methods for functional annotation including homology search to known sequence data (BLAST+/SwissProt/Uniref90), protein domain identification (HMMER/PFAM), protein signal peptide and transmembrane domain prediction (signalP/tmHMM), and comparison to currently curated annotation databases (EMBL Uniprot eggNOG/GO Pathways databases). 

The functional annotation data derived from the analysis of transcripts is integrated into a SQLite database and a whole annotation report is available for the Trinity assembly.

[The full trinotate annotation report is available here](trinotate_annotation_report.zip)
