[TOC]


RNA-Seq De Novo Assembly Pipeline
=================================

The standard MUGQIC RNA-Seq De Novo Assembly pipeline now has two protocols. One uses the 
[Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)
software suite to reconstruct transcriptomes from RNA-Seq data without using any reference genome or transcriptome.
The other one uses [Seq2Fun](https://www.seq2fun.ca/), a functional profiling tool which can directly perform functional quantification of RNA-seq reads
without transcriptome de novo assembly.

## Trinity protocol [default]

First reads are trimmed with [Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic)
and normalized in order to reduce memory requirement and decrease assembly runtime, using the Trinity
normalization utility inspired by the [Diginorm](http://arxiv.org/abs/1203.4802) algorithm.

Then, the transcriptome is assembled on normalized reads using the Trinity assembler. Trinity creates
a Trinity.fasta file with a list of contigs representing the transcriptome isoforms. Those transcripts
are grouped in components mostly representing genes.

Components and transcripts are functionally annotated using the [Trinotate](http://trinotate.sourceforge.net/) suite.

Gene abundance estimation for each sample has been performed using [RSEM](http://deweylab.biostat.wisc.edu/rsem/)
(RNA-Seq by Expectation-Maximization). Differential gene expression analysis is performed using
[DESeq](http://genomebiology.com/2010/11/10/R106) and [edgeR](http://bioinformatics.oxfordjournals.org/content/26/1/139/) R Bioconductor packages.

The DESeq and edgeR methods model **count data** by a negative binomial distribution. The parameters of
the distribution (mean and dispersion) are estimated from the data, i.e. from the read counts in the input files.
Both methods compute a measure of read abundance, i.e. expression level (called *base mean* or
*mean of normalized counts* in DESeq, and *concentration* in edgeR) for each gene and apply a hypothesis test
to each gene to evaluate differential expression. In particular, both methods determine a p-value and
a log2 fold change (in expression level) for each gene. The Log2 FC of EdgeR is reported in the differential gene
results file, one file per design.

The log2fold change is the logarithm (to basis 2) of the fold change condition from condition A to B
(mutation or treatment are the most common conditions). A "fold change" between conditions A and B at a gene
or transcript is normally computed as the ratio at gene or transcript of the base mean of scaled counts
for condition B to the base mean of scaled counts for condition A. Counts are scaled by a size factor in
a step called normalization (if the counts of non-differentially expressed genes in one sample are, on average,
twice as high as in another,  the size factor for the first sample should be twice that of the other sample).
Each column of the count table is then divided by the size factor for this column and the count values
are brought to a common scale, making them comparable. See the [EdgeR vignette](http://www.bioconductor.org/packages/2.12/bioc/vignettes/edgeR/inst/doc/edgeR.pdf) for additional information on normalization approaches used in the pipeline.

The differential gene analysis is followed by a Gene Ontology (GO) enrichment analysis.
This analysis use the [goseq approach](http://bioconductor.org/packages/release/bioc/html/goseq.html).
The goseq is based on the use of non-native GO terms resulting from trinotate annotations (see details in the section 5 of
[the corresponding vignette](http://bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf).

Thus a high quality contigs assembly is created by extracting all transcripts having a functionnal annotation as defined by trinotate,
the Top BLASTX hit and TmHMM annotations are used by default.

Finally, different exploratory data analysis (EDA) techniques are applied to filtered isoforms expression levels.
Main goals of expression level EDA are the detection of outliers, potential mislabeling,  to explore the homogeneity
of biological replicates and  to appreciate the global effects of the different experimental variables.

An HTML summary report is automatically generated by the pipeline. This report contains description of
the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various
Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible
for download directly through the report. The report includes also the main references of the software and
methods used during the analysis, together with the full list of parameters that have been passed
to the pipeline main script.

## Seq2Fun protocol

RNA-seq is a powerful tool to answer many biological questions. 
While the majority of RNA-seq data has been collected 
and analyzed in model organisms, it is increasingly collected in non-model organisms 
such as many species of environmental and/or economical importance, to answer some 
very basic questions, such as which genes are up- and down- regulated, which pathways 
are changed under different conditions. In most cases, they either lack of genome 
references or do not have high-quality genome, which has posed great challenge for 
RNA-seq data analysis for these organisms.


Therefore, Seq2Fun, an ultra-fast, assembly-free, all-in-one tool has been developed
based on a modern data structure full-text in minute space (FM)
index and burrow wheeler transformation (BWT), to functional quantification of RNA-seq 
reads for non-model organisms without transcriptome assembly and genome references.

For further information regarding Seq2Fun visit: https://www.seq2fun.ca/motivation.xhtml

The Seq2fun protocol starts with merging fastq files with multiple readsets. Then Seq2fun use the fastq
files to generate KO abundance table and several other files (for more info: [seq2fun output files](https://www.seq2fun.ca/manual.xhtml#sect4)) that can be used to perform downstream analysis 
on [NetworkAnalyst](https://www.networkanalyst.ca/NetworkAnalyst/uploads/TableUploadView.xhtml).
A HTML report for seq2fun analysis is generated.

Additionally differential KO analysis is performed using
[DESeq2](https://pubmed.ncbi.nlm.nih.gov/25516281/) and [edgeR](http://bioinformatics.oxfordjournals.org/content/26/1/139/) R Bioconductor packages.
on KO count files and result tables will be generated. Moreover, a pathway analysis using differential
analysis is performed using [fgsea](https://www.biorxiv.org/content/10.1101/060012v3). 

Usage
-----
```
#!text

usage: rnaseq_denovo_assembly.py [-h] [--help] [-c CONFIG [CONFIG ...]]
                                 [-s STEPS] [-o OUTPUT_DIR]
                                 [-j {pbs,batch,daemon,slurm}] [-f]
                                 [--no-json] [--report] [--clean]
                                 [-l {debug,info,warning,error,critical}]
                                 [--sanity-check]
                                 [--container {wrapper, singularity} <IMAGE PATH>]
                                 [--genpipes_file GENPIPES_FILE] [-d DESIGN]
                                 [-t {trinity, seq2fun}]
                                 [-r READSETS] [-v]

Version: 3.6.2

For more documentation, visit our website: https://bitbucket.org/mugqic/genpipes/

optional arguments:
  -h                    show this help message and exit
  --help                show detailed description of pipeline and steps
  -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                        config INI-style list of files; config parameters are
                        overwritten based on files order
  -s STEPS, --steps STEPS
                        step range e.g. '1-5', '3,6,7', '2,4-8'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        output directory (default: current)
  -j {pbs,batch,daemon,slurm}, --job-scheduler {pbs,batch,daemon,slurm}
                        job scheduler type (default: slurm)
  -f, --force           force creation of jobs even if up to date (default:
                        false)
  --no-json             do not create JSON file per analysed sample to track
                        the analysis status (default: false i.e. JSON file
                        will be created)
  --report              create 'pandoc' command to merge all job markdown
                        report files in the given step range into HTML, if
                        they exist; if --report is set, --job-scheduler,
                        --force, --clean options and job up-to-date status are
                        ignored (default: false)
  --clean               create 'rm' commands for all job removable files in
                        the given step range, if they exist; if --clean is
                        set, --job-scheduler, --force options and job up-to-
                        date status are ignored (default: false)
  -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                        log level (default: info)
  --sanity-check        run the pipeline in `sanity check mode` to verify that
                        all the input files needed for the pipeline to run are
                        available on the system (default: false)
  --container {wrapper, singularity} <IMAGE PATH>
                        Run inside a container providing a validsingularity
                        image path
  --genpipes_file GENPIPES_FILE, -g GENPIPES_FILE
                        Command file output path. This is the command used to
                        process the data, or said otherwise, this command will
                        "run the Genpipes pipeline". Will be redirected to
                        stdout if the option is not provided.
  -d DESIGN, --design DESIGN
                        design file
  -r READSETS, --readsets READSETS
                        readset file
  -t {trinity,seq2fun}, --type {trinity,seq2fun} 
  -v, --version         show the version information and exit

Steps:
```
![rnaseq_denovo_assembly trinity workflow diagram](https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_rnaseq_denovo_assembly.resized.png)
[download full-size diagram](https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_rnaseq_denovo_assembly.png)
```
------
1- picard_sam_to_fastq
2- trimmomatic
3- merge_trimmomatic_stats
4- insilico_read_normalization_readsets
5- insilico_read_normalization_all
6- trinity
7- exonerate_fastasplit
8- blastx_trinity_uniprot
9- blastx_trinity_uniprot_merge
10- transdecoder
11- hmmer
12- rnammer_transcriptome
13- blastp_transdecoder_uniprot
14- signalp
15- tmhmm
16- trinotate
17- align_and_estimate_abundance_prep_reference
18- align_and_estimate_abundance
19- gq_seq_utils_exploratory_analysis_rnaseq_denovo
20- differential_expression
21- filter_annotated_components
22- gq_seq_utils_exploratory_analysis_rnaseq_denovo_filtered
23- differential_expression_filtered

```
![rnaseq_denovo_assembly seq2fun workflow diagram](https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_rnaseq_denovo_assembly.resized.png)
[download full-size diagram](https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_rnaseq_denovo_assembly.png)
```
------
1- picard_sam_to_fastq
2- merge_fastq
3- seq2fun
4- differential_expression_seq2fun
5- pathway_enrichment_seq2fun

```
picard_sam_to_fastq
-------------------
Convert SAM/BAM files from the input readset file into FASTQ format
if FASTQ files are not already specified in the readset file. Do nothing otherwise.

trimmomatic
-----------
Raw reads quality trimming and removing of Illumina adapters is performed using [Trimmomatic](http://www.usadellab.org/cms/index.php?page=trimmomatic).
If an adapter FASTA file is specified in the config file (section 'trimmomatic', param 'adapter_fasta'),
it is used first. Else, 'Adapter1' and 'Adapter2' columns from the readset file are used to create
an adapter FASTA file, given then to Trimmomatic. For PAIRED_END readsets, readset adapters are
reversed-complemented and swapped, to match Trimmomatic Palindrome strategy. For SINGLE_END readsets,
only Adapter1 is used and left unchanged.

This step takes as input files:

1. FASTQ files from the readset file if available
2. Else, FASTQ output files from previous picard_sam_to_fastq conversion of BAM files

merge_trimmomatic_stats
-----------------------
The trim statistics per readset are merged at this step.

insilico_read_normalization_readsets
------------------------------------
Normalize each readset, using the Trinity normalization utility.

insilico_read_normalization_all
-------------------------------
Merge all normalized readsets together and normalize the result, using the Trinity normalization utility.

trinity
-------
Create a de novo assembly from normalized readsets using [Trinity](http://trinityrnaseq.sourceforge.net/).

exonerate_fastasplit
--------------------
Split the Trinity assembly FASTA into chunks for further parallel BLAST annotations.

blastx_trinity_uniprot
----------------------
Annotate Trinity FASTA chunks with Swiss-Prot and UniRef databases using [blastx](http://blast.ncbi.nlm.nih.gov/).

blastx_trinity_uniprot_merge
----------------------------
Merge blastx Swiss-Prot and UniRef chunks results.

transdecoder
------------
Identifies candidate coding regions within transcript sequences using [Transdecoder](http://transdecoder.github.io/).

hmmer
-----
Identifies protein domains using [HMMR](http://hmmer.janelia.org/).

rnammer_transcriptome
---------------------
Identify potential rRNA transcripts using [RNAmmer](http://www.cbs.dtu.dk/cgi-bin/sw_request?rnammer).

blastp_transdecoder_uniprot
---------------------------
Search Transdecoder-predicted coding regions for sequence homologies on UniProt using [blastp](http://blast.ncbi.nlm.nih.gov/).

signalp
-------
Predict signal peptides using [SignalP](http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?signalp).

tmhmm
-----
Predict transmembrane regions using [TMHMM](http://www.cbs.dtu.dk/cgi-bin/nph-sw_request?tmhmm).

trinotate
---------
Perform transcriptome functional annotation and analysis using [Trinotate](http://trinotate.sourceforge.net/).
All functional annotation data is integrated into a SQLite database and a whole annotation report is created.

align_and_estimate_abundance_prep_reference
-------------------------------------------
Index Trinity FASTA file for further abundance estimation using [Trinity align_and_estimate_abundance.pl utility](http://trinityrnaseq.sourceforge.net/analysis/abundance_estimation.html).

align_and_estimate_abundance
----------------------------
Estimate transcript abundance using [RSEM](http://deweylab.biostat.wisc.edu/rsem/) via
[Trinity align_and_estimate_abundance.pl utility](http://trinityrnaseq.sourceforge.net/analysis/abundance_estimation.html).

gq_seq_utils_exploratory_analysis_rnaseq_denovo
-----------------------------------------------
Exploratory analysis using the gqSeqUtils R package.

differential_expression
-----------------------
Performs differential gene expression analysis using [DESEQ](http://bioconductor.org/packages/release/bioc/html/DESeq.html) and [EDGER](http://www.bioconductor.org/packages/release/bioc/html/edgeR.html).
Merge the results of the analysis in a single csv file. Also, performs Gene Ontology analysis for RNA-Seq denovo Assembly using the Bioconductor's R package [goseq](http://www.bioconductor.org/packages/release/bioc/html/goseq.html).
Generates GO annotations for differential genes and isoforms expression analysis, based on associated GOTERMS generated by trinotate.

filter_annotated_components
---------------------------
Filter high quality contigs based on values in trinotate annotations. Recreate a high quality contigs fasta file and run Assembly statistics using the gqSeqUtils R package.

gq_seq_utils_exploratory_analysis_rnaseq_denovo_filtered
--------------------------------------------------------
Exploratory analysis using the gqSeqUtils R package using a subset of filtered transcripts

differential_expression_filtered
--------------------------------
Differential Expression and GOSEQ analysis based on filtered transcripts and genes

merge_fastq
----------------------
fastq files are merged if the sample has multiple readsets. create one fastq file for each sample

seq2fun
----------------------
For most non-model organisms, biological understanding of study outcomes is limited to
protein-coding genes with functional annotations such as KEGG pathways or
Gene Ontology or PANTHER classification system.
Therefore, Seq2Fun databases focused on functionally annotated genes such as KOs
largely meets the preferred needs of most scientists studying non-model organisms.

**Note:** We have stored number of pre-built databases by the Seq2Fun developers 
and the list can be found below. 

* Eukaryotes
* Animals
* Plants
* Fungi
* Protists
* Mammals
* Birds
* Reptiles
* Amphibians
* Fishes
* Arthropods
* Nematodes

For further information please visit
https://github.com/xia-lab/Seq2Fun#step-3-database-download

You can specify the required database in a custom.ini file by changing "group" key. The default
group is "birds"

* *These KOs in the database are KOs assigned to KEGG pathways and they are only
a proportion of whole list of KOs.*
* *All KOs include KOs not assigned to KEGG pathways.*

This step takes as input files:

 1. FASTQ files from the readset file if available
 2. Else, FASTQ output files from previous picard_sam_to_fastq conversion of BAM files

 This step perform seq2fun analysis and generates [output files](https://www.seq2fun.ca/manual.xhtml#sect4) including KO abundance table and [KO mapped fastq files](https://www.seq2fun.ca/manual.xhtml#sect20).
These output files can be used for downstream analysis using
[Network Analyst](https://www.networkanalyst.ca/NetworkAnalyst/uploads/TableUploadView.xhtml)
 web application The interested comparisons should be specified using a design file (Format is similar to the 
default RNA-seq design file). Therefore, only pairwise comparisons
are possible unlike the original Seq2Fun program (treatment and controls will be added according to the 1 and 2 in the design file)
If you want to compare multiple groups you will have to break the comparison in to pairs.

E.g for 3 group comparison design file should be as follows

    Sample  A_vs_B   B_vs_C   A_vs_C
     A1       1        0       1
     B1       2        1       0
     A2       1        0       1
     B2       2        1       0
     C1       0        2       2
     C2       0        2       2


Seq2Fun analysis is performed using fastq files and it generates a 
sample table.

differential_expression_seq2fun
----------------------

DESE2 and EdgeR will be used to perform differential KO expression analysis

pathway_enrichment_seq2fun
----------------------

A [KEGG ortholog](https://www.genome.jp/kegg/ko.html) pathway analysis will be performed using fgsea R package

Current KEGG database used for the analysis was created on August, 2021. 
If you'd like to update the database,
please follow below instructions to download the latest KEGG pathway database and create
pathway list for the kegg_all key in the ini file and kegg.rds
Download the latest KEGG database

**In linux**

```
wget http://rest.kegg.jp/list/pathway`
sed -i 's/path://g' pathway
mv pathway KEGG_all_pathways.txt
```

**in R**

```
library(data.table)
pathway_list <- fread("KEGG_all_pathways.txt", header =F)`
kegg_ko <- lapply(unique(pathway_list$mapID), fun)
names(kegg_ko) <- (pathway_list$mapID)
kegg <- lapply(kegg_ko, substring, 4)
saveRDS(kegg, "kegg.rds")
```

Next, specify created files in a custom ini file.

**user_pathway_list** is a file with pathway that are interested with KEGG map id 
(only one column). You must specify the file path to the user_pathway_list in a custom.ini

